# Data normalisation: centering, scaling, quantile normalisation {#sec:norm}

**Learning objectives**

- What is data normalisation and why do we need it?
- What are centering and scaling?
- What is quantile normalisation?
- How to assess if normalisation worked.

## Introduction

In a typical high throughput experiment, we assay thousands of
features (gene transcripts, proteins, metabolites, ...) is a certain
number of biologically diverse samples (from about 6 to hundreds of
thousands).

But what do you think would happen is we took the same sample, such as
a suspension of U2OS cells (a cell line stemming from bone cancer),
split it in two, and measured it twice using the exact same protocol?
Will the measurements be exactly identical? Why? These would be
**technical replicates**.

If we now take two independent different cell cultures of the same
cell line, do you expect there to be larger, similar or smaller
difference than in the examples above? One would refer to **biological
replicates** in this case.

- See technical replication in RNA-Seq paper from Irizayry group.
- Find some data with tech and bio reps.

Experiments aim at measuring biological variability by comparing
biological conditions in a controlled setting. The be able to measure
any biologically signal in the data, the biological variability must
be larger than the technical variability.

## Data transformation

## Centring and scaling


### Generating data


`r msmbstyle::question_begin()`
We are going to start by generating a data set to precisely illustrate
the effect of the methods.

1. Use the `rnorm()` function to generate a distrubition of 1000
   values centred around 0 and with a standard deviation
   of 2. Visualise these data.

2. Generate four such distribution with parameters *N(6, 2)*,
   *N(4,2)*, *N(4, 1)*, *N(7, 3)* and create a matrix or dataframe
   with rownames `gene1` to `gene1000` and colnames `sample1` to
   `sample4`. Visualise these data and disucss whether these samples
   could be compared against each other. Do assure replication of this
   simulation, set the random number generation seed to 123.
`r msmbstyle::question_end()`


`r msmbstyle::solution_begin()`
```{r sln1}
n <- 1000
set.seed(123)
x <- data.frame(sample1 = rnorm(n, 6, 2),
                sample2 = rnorm(n, 4, 2),
                sample3 = rnorm(n, 4, 1),
                sample4 = rnorm(n, 7, 3))
head(x)
boxplot(x)
```
`r msmbstyle::solution_end()`


Centering refers to the action of modifying the mean value of a set of
values by subtracting a fixed value from each indivual value. On the
figure above, this equates to shifting the values up to down. A
typical value is the mean of all the data to be centred.

`r msmbstyle::question_begin()`
For each column, calculate its mean value and subtract it from the
values. Visualise and interprete the centred data.
`r msmbstyle::question_end()`

`r msmbstyle::solution_begin()`
```{r}
cmns <- colMeans(x)
for (i in 1:ncol(x))
    x[, i] <- x[, i] - cmns[i]
boxplot(x)
```
`r msmbstyle::solution_end()`
