# Hypothesis testing {#sec:testing}

Prerequisites, see WFARM1247 (Traitement statistique des donn√©es)

## Refresher



- Tossing a coin.
- Binomial distribution
- p-value 

## A biological example

We are going to use the `tdata1` dataset from the `rWSBIM1322` package
that provide gene expression data for 100 genes and 6 samples, three
in group A and 3 in group B.

```{r}
library("rWSBIM1322")
data(tdata2)
head(tdata1)
```

`r msmbstyle::question_begin()`
Visualise the distribution of the `tdata1` data and, if necessary,
log-transform it.
`r msmbstyle::question_end()`

`r msmbstyle::solution_begin()`
```{r tex1}
log_tdata1 <- log2(tdata1)
par(mfrow = c(1, 2))
limma::plotDensities(tdata1)
limma::plotDensities(log_tdata1)
```
`r msmbstyle::solution_end()`


We are now going to apply a t-test to feature 73, comparing the
expression intensities in groups A and B. In R, this can be done with
the `t.test` function:

```{r}
x <- log_tdata1[73, ]
t.test(x[1:3], x[4:6])
```

`r msmbstyle::question_begin()`
- Interpret the results of the test above.
- Repeat it on another features.
`r msmbstyle::question_end()`


`r msmbstyle::question_begin()`
We would now like to repeat the same analysis on the 100 genes. 
- Write a function that will take a vector as input and perform a
  t-test of the first values (our group A) against the 3 last values
  (our group B) and returns the p-values.
- Apply the test to all the genes.
- How many significantly differentically expressed genes do you find?
  What features are of possible biological interest?
`r msmbstyle::question_end()`


```{r}
my_t_test <- function(x) {
    t.test(x[1:3], x[4:6])$p.value
}
pvals <- apply(log_tdata1, 1, my_t_test)
table(pvals < 0.05)
head(sort(pvals))
```

`r msmbstyle::question_begin()`
The data above have been generated with the `rnorm` function for all
samples. 
- Do you still think any of the features show significant differences?
- Why are there still some features (around 5%) that show a
  significant p-value at an alpha of 0.05?
`r msmbstyle::question_end()`




```{r}
hist(pvals)
```

## Adjustment for multiple testing

Given the multitue of test, the type I error is not controlled
anymore. We need to take this into account and adjust the p-values for
multiple testing. Below we apply the Benjamini-Hochberg procedure
using the `p.adjust` function and confirm that none of the features
are differentially expressed.

```{r}
adj_pvals <- p.adjust(pvals, method = "BH")
any(adj_pvals < 0.05)
min(adj_pvals)
```

## Linear regression

## Empirical approximations 

- Bootstrapping to generate null distributions

## A word of caution

- p-hacking and p-harking


## Additional exercises

`r msmbstyle::question_begin()`
Generate random data using `rnorm` for 100 genes and 6 samples and
test for differential expression, comparing for each gene the 3 first
samples against the 3 last samples. Verify that you identify about 5
p-values smaller that 0.05. Visualise and interprete the histogramme
of these p-values.
`r msmbstyle::question_end()`
