<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 7 Hypothesis testing | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-10-27" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 7 Hypothesis testing | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a>
<a href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a id="active-page" href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a><ul class="toc-sections">
<li class="toc"><a href="#refresher"> Refresher</a></li>
<li class="toc"><a href="#a-biological-example"> A biological example</a></li>
<li class="toc"><a href="#a-more-realistic-biological-example"> A more realistic biological example</a></li>
<li class="toc"><a href="#adjustment-for-multiple-testing"> Adjustment for multiple testing</a></li>
<li class="toc"><a href="#linear-regression"> Linear regression</a></li>
<li class="toc"><a href="#empirical-approximations"> Empirical approximations</a></li>
<li class="toc"><a href="#a-word-of-caution"> A word of caution</a></li>
<li class="toc"><a href="#additional-exercises-5"> Additional exercises</a></li>
</ul>
<a href="sec-ul.html"><span class="toc-section-number">8</span> Unsupervised learning</a>
<a href="sec-dimred.html"><span class="toc-section-number">9</span> Dimensionality reduction</a>
<a href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:testing" class="section level1">
<h1>
<span class="header-section-number">Chapter 7</span> Hypothesis testing</h1>
<p>For prerequisites within the Biomedical sciences masters degree at the UCLouvain, see <em>WFARM1247</em> (Traitement statistique des données).</p>
<div id="refresher" class="section level2">
<h2>
<span class="header-section-number">7.1</span> Refresher</h2>
<p>We have flipped a coing 100 times and have obtained the following results.</p>
<pre><code>## flips
##  H  T 
## 62 38</code></pre>
<p>If the coin were unbiased we expect roughly 50 heads. Can we make any claims regarding the biaised or unbiaised nature of that coin?</p>
<p>A coin toss can be modeled by a binomial distribution. The histogram below shows the binomial statistic for 0 to 100 heads; it represents the binomial density of for an unbiased coin. The vertical line shows the number of head observed.</p>
<div class="figure">
<span id="fig:tosshist"></span>
<p class="caption marginnote shownote">
Figure 7.1: Binomial density of for an unbiased coin to get 0 to 100 heads. The full area of the histogram sums to 1.
</p>
<img src="WSBIM1322_files/figure-html/tosshist-1.png" alt="Binomial density of for an unbiased coin to get 0 to 100 heads. The full area of the histogram sums to 1." width="672">
</div>
<p>Above, we see that the most likely outcome would be 50 heads with a probability of 0.0795892. No head and 100 heads have respectively 7.888609110^{-31} and 7.888609110^{-31} probability.</p>
<p>We set</p>
<ul>
<li>
<span class="math inline">\(H_0\)</span>: the coin is fair</li>
<li>
<span class="math inline">\(H_1\)</span>: the coin is baised</li>
</ul>
<p>If 62 isn’t deemed too <em>extreme</em>, then we won’t reject <span class="math inline">\(H_0\)</span> and conclude that the coin in faire. If 62, then we reject <span class="math inline">\(H_0\)</span> and accept <span class="math inline">\(H_1\)</span>, and conclude that the coin is baised.</p>
<p>To define <em>extreme</em>, we set <span class="math inline">\(\alpha = 0.05\)</span> and rejet <span class="math inline">\(H_0\)</span> if our result is outside of the 95% most probably values.</p>
<div class="figure">
<span id="fig:tosshist2"></span>
<p class="caption marginnote shownote">
Figure 7.2: Binomial density of for an unbiased coin to get 0 to 100 heads. The areas in red sum to 0.05.
</p>
<img src="WSBIM1322_files/figure-html/tosshist2-1.png" alt="Binomial density of for an unbiased coin to get 0 to 100 heads. The areas in red sum to 0.05." width="672">
</div>
<p>We can also compute the p-value, the tells us how likely we are to see such a extreme or more extreme value under <span class="math inline">\(H_0\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">binom.test</span>(<span class="dt">x =</span> <span class="dv">62</span>, <span class="dt">n =</span> <span class="dv">100</span>, <span class="dt">p =</span> <span class="fl">0.5</span>)</code></pre></div>
<pre><code>## 
##  Exact binomial test
## 
## data:  62 and 100
## number of successes = 62, number of trials = 100, p-value =
## 0.02098
## alternative hypothesis: true probability of success is not equal to 0.5
## 95 percent confidence interval:
##  0.5174607 0.7152325
## sample estimates:
## probability of success 
##                   0.62</code></pre>
<p>Whenever we make such a decision, we will be in one of the following situations:</p>
<table>
<thead><tr class="header">
<th></th>
<th>
<span class="math inline">\(H_0\)</span> is true</th>
<th>
<span class="math inline">\(H_1\)</span> is true</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Reject <span class="math inline">\(H_0\)</span>
</td>
<td>Type I (false positive)</td>
<td>True positive</td>
</tr>
<tr class="even">
<td>Do not reject <span class="math inline">\(H_0\)</span>
</td>
<td>True negative</td>
<td>Type II (false negative)</td>
</tr>
</tbody>
</table>
<p>See below for a step by step guide to this example.</p>
</div>
<div id="a-biological-example" class="section level2">
<h2>
<span class="header-section-number">7.2</span> A biological example</h2>
<p>A typical biological example consists in measuring a gene of interest in two populations of interest (say control and group), represented by biological replicates. The figure below represents the distribution of the gene of interest in the two populations, with the expression intensities of triplicates in each population.</p>
<div class="figure">
<span id="fig:exdists"></span>
<p class="caption marginnote shownote">
Figure 7.3: Expression of a gene in two populations with randomly chosen triplicates.
</p>
<img src="WSBIM1322_files/figure-html/exdists-1.png" alt="Expression of a gene in two populations with randomly chosen triplicates." width="672">
</div>
<p>We don’t have access to the whole population and thus use a sample thereof (the replicated measurements) to estimate the population parameters. We have</p>
<table>
<thead><tr class="header">
<th></th>
<th align="right">control</th>
<th align="right">group</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>rep.1</td>
<td align="right">5.327429</td>
<td align="right">11.10576</td>
</tr>
<tr class="even">
<td>rep.2</td>
<td align="right">5.723787</td>
<td align="right">11.19393</td>
</tr>
<tr class="odd">
<td>rep.3</td>
<td align="right">7.870450</td>
<td align="right">13.57260</td>
</tr>
<tr class="even">
<td>mean</td>
<td align="right">6.307222</td>
<td align="right">11.95743</td>
</tr>
<tr class="odd">
<td>sd</td>
<td align="right">1.368224</td>
<td align="right">1.39947</td>
</tr>
</tbody>
</table>
<p>We set our hypothesis as</p>
<ul>
<li>
<span class="math inline">\(H_0\)</span>: the means of the two groups are the same.</li>
<li>
<span class="math inline">\(H_1\)</span>: the means of the two groups are different.</li>
</ul>
<p>and calculate a two-sided, two-sample t-test (assuming unequal variances) with</p>
<p><span class="math display">\[
t = \frac{ \bar X_{1} - \bar X_{2} }
         { \sqrt{ \frac{ s_{1}^{2} }{ N_{1} } \frac{ s_{2}^{2} }{ N_{2} } } }
\]</span></p>
<p>where <span class="math inline">\(\bar X_i\)</span>, <span class="math inline">\(s_{i}^{2}\)</span> and <span class="math inline">\(N_i\)</span> are the mean, variance and size of samples 1 and 2. In R, we do this with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(s1, s2)</code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  s1 and s2
## t = -5.0003, df = 3.998, p-value = 0.007499
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -8.788162 -2.512255
## sample estimates:
## mean of x mean of y 
##  6.307222 11.957431</code></pre>
<p>Note that in practice, we would apply moderated versions of the t-test, such as the one provided in the <em><a href="https://bioconductor.org/packages/3.9/limma">limma</a></em> package and also widely applied to RNA-Seq count data.</p>
</div>
<div id="a-more-realistic-biological-example" class="section level2">
<h2>
<span class="header-section-number">7.3</span> A more realistic biological example</h2>
<p>Let’s now use the <code>tdata1</code> dataset from the <code>rWSBIM1322</code> package that provide gene expression data for 100 genes and 6 samples, three in group A and 3 in group B.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"rWSBIM1322"</span>)
<span class="kw">data</span>(tdata1)
<span class="kw">head</span>(tdata1)</code></pre></div>
<pre><code>##                 A1         A2         A3         B1         B2        B3
## feature1  429.6721   433.3133  1806.2773  3534.6396  4541.2757 1139.8326
## feature2 1320.8827  1085.5660 10643.9541   239.7528 14177.4412  678.5271
## feature3  321.5151   289.6480  9236.7965 15745.7491   443.8737  198.5716
## feature4 9348.7749  1274.8021   647.2523   601.6313   595.6266 1040.1568
## feature5 1616.9009   413.2385    43.0974 10143.6595   575.0549 4048.5851
## feature6  328.3438 11866.2342 32661.9472  8331.8745   608.3194 9331.7692</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise the distribution of the <code>tdata1</code> data and, if necessary, log-transform it.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-34" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-34', 'sol-start-34')"></span>
</p>
<div id="sol-body-34" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">log_tdata1 &lt;-<span class="st"> </span><span class="kw">log2</span>(tdata1)
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
limma<span class="op">::</span><span class="kw">plotDensities</span>(tdata1)
limma<span class="op">::</span><span class="kw">plotDensities</span>(log_tdata1)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/tex1-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>We are now going to apply a t-test to feature 73, comparing the expression intensities in groups A and B. In R, this can be done with the <code>t.test</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>log_tdata1[<span class="dv">73</span>, ]
<span class="kw">t.test</span>(x[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>], x[<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>])</code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x[1:3] and x[4:6]
## t = 0.1728, df = 3.999, p-value = 0.8712
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -3.444807  3.902010
## sample estimates:
## mean of x mean of y 
##  10.14391   9.91531</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Interpret the results of the test above.</li>
<li>Repeat it with another features.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>We would now like to repeat the same analysis on the 100 genes.</p>
<ul>
<li><p>Write a function that will take a vector as input and perform a t-test of the first values (our group A) against the 3 last values (our group B) and returns the p-values.</p></li>
<li><p>Apply the test to all the genes.</p></li>
<li><p>How many significantly differentically expressed genes do you find? What features are of possible biological interest?</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-35" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-35', 'sol-start-35')"></span>
</p>
<div id="sol-body-35" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_t_test &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
    <span class="kw">t.test</span>(x[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>], x[<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>])<span class="op">$</span>p.value
}
pvals &lt;-<span class="st"> </span><span class="kw">apply</span>(log_tdata1, <span class="dv">1</span>, my_t_test)
<span class="kw">table</span>(pvals <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##    98     2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">sort</span>(pvals))</code></pre></div>
<pre><code>##  feature19  feature61  feature20  feature84   feature1   feature7 
## 0.02967230 0.03595002 0.08576252 0.08958506 0.10645476 0.10832969</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>The data above have been generated with the <code>rnorm</code> function for all samples. - Do you still think any of the features show significant differences? - Why are there still some features (around 5%) that show a significant p-value at an alpha of 0.05?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>To answer these questions, let’s visualise the calculated p-values with a histogramme.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(pvals)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
</div>
<div id="adjustment-for-multiple-testing" class="section level2">
<h2>
<span class="header-section-number">7.4</span> Adjustment for multiple testing</h2>
<p>Given the multitue of test, the type I error is not controlled anymore. We need to take this into account and adjust the p-values for multiple testing. Below we apply the Benjamini-Hochberg procedure using the <code>p.adjust</code> function and confirm that none of the features are differentially expressed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">adj_pvals &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(pvals, <span class="dt">method =</span> <span class="st">"BH"</span>)
<span class="kw">any</span>(adj_pvals <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">min</span>(adj_pvals)</code></pre></div>
<pre><code>## [1] 0.899736</code></pre>
</div>
<div id="linear-regression" class="section level2">
<h2>
<span class="header-section-number">7.5</span> Linear regression</h2>
</div>
<div id="empirical-approximations" class="section level2">
<h2>
<span class="header-section-number">7.6</span> Empirical approximations</h2>
<ul>
<li>Bootstrapping to generate null distributions</li>
</ul>
</div>
<div id="a-word-of-caution" class="section level2">
<h2>
<span class="header-section-number">7.7</span> A word of caution</h2>
<ul>
<li>p-hacking and p-harking</li>
</ul>
</div>
<div id="additional-exercises-5" class="section level2">
<h2>
<span class="header-section-number">7.8</span> Additional exercises</h2>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Generate random data using <code>rnorm</code> for 100 genes and 6 samples and test for differential expression, comparing for each gene the 3 first samples against the 3 last samples. Verify that you identify about 5 p-values smaller that 0.05. Visualise and interprete the histogramme of these p-values.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div id="the-coin-example-step-by-step" class="section level3 unnumbered">
<h3>The coin example, step by step</h3>
<blockquote>
<p>This step by step example is taken from Oliver Crook’s intro lecture on statistics from the <a href="https://uclouvain-cbio.github.io/BSS2019/">2019 Bioinformatics Summer School</a> at the UCLouvain.</p>
</blockquote>
<p>The following code chunk simulates 100 coin toss from a biased coin.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">2</span>) 
n &lt;-<span class="st"> </span><span class="dv">100</span>
p &lt;-<span class="st"> </span><span class="fl">0.59</span>
flips &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"H"</span>, <span class="st">"T"</span>), <span class="dt">size =</span> n,
                <span class="dt">replace =</span> <span class="ot">TRUE</span>,
                <span class="dt">prob =</span> <span class="kw">c</span>(p, <span class="dv">1</span> <span class="op">-</span><span class="st"> </span>p))</code></pre></div>
<p>If the coin were unbiased we expect roughly 50 heads. Let us see how many heads and tails there are.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(flips)</code></pre></div>
<pre><code>## flips
##  H  T 
## 62 38</code></pre>
<p>We calculate the binomial statistic for a number of flips between 0 and 100. This is the binomial density for an unbiased coin.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"dplyr"</span>)
num_heads &lt;-<span class="st"> </span><span class="kw">sum</span>(flips <span class="op">==</span><span class="st"> "H"</span>)
binomial_dens &lt;-
<span class="st">    </span><span class="kw">tibble</span>(<span class="dt">k =</span> <span class="dv">0</span><span class="op">:</span>n) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">p =</span> <span class="kw">dbinom</span>(k, <span class="dt">size =</span> n, <span class="dt">prob =</span> <span class="fl">0.5</span>))</code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Write some code to check that the probabilties from the binomial statistic sum to <span class="math inline">\(1\)</span>.</p></li>
<li><p>Change the <code>prob</code> argument and show that the probabilities still sum to <span class="math inline">\(1\)</span>.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The following code chunk plots the binomial statistic and the number of heads observed is marked in blue.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"ggplot2"</span>)
<span class="kw">ggplot</span>(binomial_dens, <span class="kw">aes</span>(<span class="dt">x =</span> k, <span class="dt">y =</span> p)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> num_heads)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-36-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Change the prob argument above and re-plot the binomial statistic, what do you notice about how the distribution is centered?</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Now, we set the size of the reject threshold, this is a choice and corresponds to how many false discoveries we are happy to allow.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha &lt;-<span class="st"> </span><span class="fl">0.05</span></code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Without looking below use the arrange function from <code>dplyr</code> to order the probabilities, smallest first.</p></li>
<li><p>Looking at the output, what is the most unlikely number of heads to observe?</p></li>
<li><p>Looking at the output, what is the most likely number of heads to observe?</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-36" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-36', 'sol-start-36')"></span>
</p>
<div id="sol-body-36" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">binomial_dens <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(p <span class="op">==</span><span class="st"> </span><span class="kw">max</span>(p))</code></pre></div>
<pre><code>## # A tibble: 1 x 2
##       k      p
##   &lt;int&gt;  &lt;dbl&gt;
## 1    50 0.0796</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">binomial_dens <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(p <span class="op">==</span><span class="st"> </span><span class="kw">min</span>(p))</code></pre></div>
<pre><code>## # A tibble: 2 x 2
##       k        p
##   &lt;int&gt;    &lt;dbl&gt;
## 1     0 7.89e-31
## 2   100 7.89e-31</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>What does the following code chunk do?</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">binomial_dens &lt;-
<span class="st">    </span><span class="kw">arrange</span>(binomial_dens, p) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">reject =</span> (<span class="kw">cumsum</span>(p) <span class="op">&lt;=</span><span class="st"> </span>alpha))</code></pre></div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Let us plot the reject region in red</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(binomial_dens) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="kw">aes</span>(<span class="dt">x =</span> k, <span class="dt">y =</span> p, <span class="dt">fill =</span> reject), <span class="dt">stat =</span> <span class="st">"identity"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_manual</span>(
        <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">`</span><span class="dt">TRUE</span><span class="st">`</span> =<span class="st"> "red"</span>, <span class="st">`</span><span class="dt">FALSE</span><span class="st">`</span> =<span class="st"> "#5a5a5a"</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> num_heads, <span class="dt">col =</span> <span class="st">"blue"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">"none"</span>)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-40-1.png" width="672"></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Is there evidence that our coin is biased?</li>
<li>Change the size of the reject region to a smaller value, what is our conclusion now?</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The above test already has an easy to use funciton in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">binom.test</span>(<span class="dt">x =</span> num_heads, <span class="dt">n =</span> n, <span class="dt">p =</span> <span class="fl">0.5</span>)</code></pre></div>
<pre><code>## 
##  Exact binomial test
## 
## data:  num_heads and n
## number of successes = 62, number of trials = 100, p-value =
## 0.02098
## alternative hypothesis: true probability of success is not equal to 0.5
## 95 percent confidence interval:
##  0.5174607 0.7152325
## sample estimates:
## probability of success 
##                   0.62</code></pre>

</div>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-mlintro.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-ul.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-10-27
</p>
</div>
</div>



</body>
</html>
