<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 12 Conclusions | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2020-01-08" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 12 Conclusions | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a>
<a href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-dimred.html"><span class="toc-section-number">8</span> Unsupervised learning: dimensionality reduction</a>
<a href="sec-ul.html"><span class="toc-section-number">9</span> Unsupervised learning: clustering</a>
<a href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a id="active-page" href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a><ul class="toc-sections">
<li class="toc"><a href="#additional-exercise"> Additional exercise</a></li>
</ul>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:ccl" class="section level1">
<h1>
<span class="header-section-number">Chapter 12</span> Conclusions</h1>
<p>In this course, we have consolidated our skills in data analysis and visualisation using R. In particular, this course has focused on interpretation and understanding of the outputs.</p>
<p>We have also learned and applied new tools, in particular how to manipulate sequence data using <code>Biostrings</code> and how to use statistical learning tools to explore data and identify patterns of biological interest. In terms of statistical testing and machine learning, it can be useful to summarise the different classes of techniques we have touch on using the figures below. The grid on each of these figures represents a matrix with quantitative values with features (genes, proteins, transcripts, …) along the rows and samples along the columns. Annotations of features or samples are presented as coloured boxes on the right or top if the matrices.</p>
<p>In hypothesis test, we start with quantitative data and an experimental design, i.e. sample annotation that group samples in biologically relevant groups. The output of hypothesis testing is a set of metrics for each features (a p-value, an adjusted p-value, a fold-change, …) that informs whether that feature shows any difference between the biological groups of interest.</p>
<div class="figure">
<span id="fig:unnamed-chunk-100"></span>
<p class="caption marginnote shownote">
Figure 12.1: Hypothesis testing.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-100-1.png" alt="Hypothesis testing." width="672">
</div>
<p>When performing clustering (unsupervised machine learning), we only have our quantitative data as input, and the clustering algorithm (whether k-means, hierarchical clustering among many others) suggests a set of groups to cluster the features (as shown below) or samples.</p>
<div class="figure">
<span id="fig:unnamed-chunk-101"></span>
<p class="caption marginnote shownote">
Figure 12.2: Clustering of features.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-101-1.png" alt="Clustering of features." width="672">
</div>
<p>When performing dimensionality reduction with, for example PCA, one starts with a <em>n</em> by <em>m</em> data set as input to reduce the number of dimensions in either direction. On the figure below, the number of features <em>n</em> was reduced to 2 to, typically, visualise the sample along a scatter plot.</p>
<div class="figure">
<span id="fig:unnamed-chunk-102"></span>
<p class="caption marginnote shownote">
Figure 12.3: Dimensionality reduction.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-102-1.png" alt="Dimensionality reduction." width="672">
</div>
<p>In classification (supervised machine learning), we need labelled data, i.e. a set of sample (top on the figures below) or features (bottom on the figure below). The classifier uses the data to learn to assign labels and applies that learnt model it infer the most likely label of the unlabelled data.</p>
<div class="figure">
<span id="fig:unnamed-chunk-103"></span>
<p class="caption marginnote shownote">
Figure 12.4: Classification of samples (top) or features (bottom).
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-103-1.png" alt="Classification of samples (top) or features (bottom)." width="672">
</div>
<p>The next steps of the curriculum (course <a href="https://uclouvain-cbio.github.io/WSBIM2122/%5D">WSBIM2122</a>) will build upon the skills gained in this course to fully analyse complete datasets from omics technologies, using state-of-the-art statistical and machine learning methods and software. The course will be project based: each experiment and associated technologies will be introduced, the analysis pipeline will be explained, and the students will then implement and present the data analysis and critical interpretation of the results.</p>
<div id="additional-exercise" class="section level2">
<h2>
<span class="header-section-number">12.1</span> Additional exercise</h2>
<p>This final exercise integrates several techniques seen throughout this course.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>The the RNA-Seq data and sample annotation in the files returned by the <code>rWSBIM1207::kem2.tsv()</code> function. How many genes have been assayed? Describe the experimental design at hand.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-79" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-79', 'sol-start-79')"></span>
</p>
<div id="sol-body-79" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">read.delim</span>(rWSBIM1207<span class="op">::</span><span class="kw">kem2.tsv</span>()[<span class="dv">1</span>],
                          <span class="dt">row.names =</span> <span class="dv">1</span>))
<span class="kw">dim</span>(e)</code></pre></div>
<pre><code>## [1] 4774   16</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Below, we harmonise the samples names to have dots in the
## experimental design dataframe (pd: phenotypic data) and expression
## matix.
pd &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_tsv</span>(rWSBIM1207<span class="op">::</span><span class="kw">kem2.tsv</span>()[<span class="dv">2</span>]) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">sample_id =</span> <span class="kw">sub</span>(<span class="st">"-"</span>, <span class="st">"."</span>, sample_id)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">group =</span> <span class="kw">paste</span>(cell_type, treatment, <span class="dt">sep =</span> <span class="st">"."</span>))</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   sample_id = col_character(),
##   jurkat = col_character(),
##   cell_type = col_character(),
##   treatment = col_character()
## )</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pd</code></pre></div>
<pre><code>## # A tibble: 16 x 5
##    sample_id jurkat cell_type treatment  group       
##    &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;       
##  1 KEM182.01 yes    A         none       A.none      
##  2 KEM182.02 yes    A         none       A.none      
##  3 KEM182.03 yes    A         none       A.none      
##  4 KEM182.04 yes    A         none       A.none      
##  5 KEM182.05 yes    B         none       B.none      
##  6 KEM182.06 yes    B         none       B.none      
##  7 KEM182.07 yes    B         none       B.none      
##  8 KEM182.08 yes    B         none       B.none      
##  9 KEM182.09 yes    A         stimulated A.stimulated
## 10 KEM182.10 yes    A         stimulated A.stimulated
## 11 KEM182.11 yes    A         stimulated A.stimulated
## 12 KEM182.12 yes    A         stimulated A.stimulated
## 13 KEM182.13 yes    B         stimulated B.stimulated
## 14 KEM182.14 yes    B         stimulated B.stimulated
## 15 KEM182.15 yes    B         stimulated B.stimulated
## 16 KEM182.16 yes    B         stimulated B.stimulated</code></pre>
<p>Cell types A and B have been assay in a stimulted and non-stimulated (none) conditions. There are 4 replicates in each of the 4 groups.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Log-2 transform the expression data and visualise the distributions of the expression values in each samples.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-80" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-80', 'sol-start-80')"></span>
</p>
<div id="sol-body-80" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">log2</span>(e <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">boxplot</span>(e, <span class="dt">col =</span> <span class="kw">as.factor</span>(pd<span class="op">$</span>group))</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-106"></span>
<p class="caption marginnote shownote">
Figure 12.5: Expression values in cell types A and B without (black and gree) and with stimulation (red and blue).
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-106-1.png" alt="Expression values in cell types A and B without (black and gree) and with stimulation (red and blue)." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Continue without preforming any normalisation.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Removes genes that aren’t expressed in any samples, i.e. those that have a sum of expression equal to 0 accross all samples. How many genes unexpressed genes are there?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-81" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-81', 'sol-start-81')"></span>
</p>
<div id="sol-body-81" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">rowSums</span>(e) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</code></pre></div>
<pre><code>## 
## TRUE 
## 4774</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span>e[<span class="kw">rowSums</span>(e) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, ]</code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise and interpret the experiment desing (cell type and treatment) using a principal component analyis.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-82" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-82', 'sol-start-82')"></span>
</p>
<div id="sol-body-82" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"factoextra"</span>)
pca_samples  &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(e), <span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)
<span class="kw">fviz_pca_ind</span>(pca_samples, <span class="dt">habillage =</span> pd<span class="op">$</span>group)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-108"></span>
<p class="caption marginnote shownote">
Figure 12.6: PCA of <em>kem2</em> samples.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-108-1.png" alt="PCA of *kem2* samples." width="672">
</div>
<p>The PCA shows that samples group by treatment along the second PC (22.3% of variance explained) and that close to 50% of the variance is explained by variability along all samples.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use a t test to identify the genes that are differentially expressed between the stimulated and non-stimulated samples of cell type A.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-83" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-83', 'sol-start-83')"></span>
</p>
<div id="sol-body-83" class="solution-body" style="display: none;">
<p>We start by selecting the relevant samples and create a new expression matrix matching cell type A. Visualise the results on a volcano plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cell_type_a &lt;-<span class="st"> </span>pd <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(cell_type <span class="op">==</span><span class="st"> "A"</span>)
cell_type_a</code></pre></div>
<pre><code>## # A tibble: 8 x 5
##   sample_id jurkat cell_type treatment  group       
##   &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;       
## 1 KEM182.01 yes    A         none       A.none      
## 2 KEM182.02 yes    A         none       A.none      
## 3 KEM182.03 yes    A         none       A.none      
## 4 KEM182.04 yes    A         none       A.none      
## 5 KEM182.09 yes    A         stimulated A.stimulated
## 6 KEM182.10 yes    A         stimulated A.stimulated
## 7 KEM182.11 yes    A         stimulated A.stimulated
## 8 KEM182.12 yes    A         stimulated A.stimulated</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e_a &lt;-<span class="st"> </span>e[, cell_type_a[[<span class="dv">1</span>]]]
e_a[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</code></pre></div>
<pre><code>##                 KEM182.01 KEM182.02 KEM182.03 KEM182.04
## ENSG00000000419  9.491853  8.997179  9.306062  9.262095
## ENSG00000001084  9.888743  9.519636  9.717676  9.729621
## ENSG00000001167 11.285980 10.980140 11.114393 11.249706
## ENSG00000001461  9.214319  8.353147  8.781360  9.172428
## ENSG00000001497 10.343186  9.885696 10.048487 10.284246
## ENSG00000001629  9.927778  9.372865  9.554589  9.573647
## ENSG00000002549  9.805744  9.527477  9.734710  9.636625
## ENSG00000002586 11.982281 11.448116 11.897845 11.936638
## ENSG00000002834 11.558899 10.872675 11.136350 11.396605
## ENSG00000003056 10.803324 10.482808 10.663558 10.736402</code></pre>
<p>Below, we perform a t test on each of the 4774 genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_t_test &lt;-<span class="st"> </span><span class="cf">function</span>(x)
    <span class="kw">t.test</span>(x[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>], x[<span class="dv">5</span><span class="op">:</span><span class="dv">8</span>])<span class="op">$</span>p.value

res &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">rownames</span>(e))
res<span class="op">$</span>pv &lt;-<span class="st"> </span><span class="kw">apply</span>(e_a, <span class="dv">1</span>, my_t_test)
<span class="kw">hist</span>(res<span class="op">$</span>pv)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-110"></span>
<p class="caption marginnote shownote">
Figure 12.7: Histogram of p-values, showing an enrichment of small values.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-110-1.png" alt="Histogram of p-values, showing an enrichment of small values." width="672">
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res<span class="op">$</span>adjp &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(res<span class="op">$</span>pv, <span class="dt">method =</span> <span class="st">"BH"</span>)

my_lfc &lt;-<span class="st"> </span><span class="cf">function</span>(x)
    <span class="kw">mean</span>(x[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]) <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(x[<span class="dv">5</span><span class="op">:</span><span class="dv">8</span>])
res<span class="op">$</span>lfc &lt;-<span class="st"> </span><span class="kw">apply</span>(e_a, <span class="dv">1</span>, my_lfc)

res <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">arrange</span>(adjp) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">head</span>()</code></pre></div>
<pre><code>##             pv         adjp        lfc
## 1 6.937532e-09 2.734553e-05  -6.698396
## 2 1.146323e-08 2.734553e-05  -9.952726
## 3 4.751844e-08 5.636475e-05  -9.303326
## 4 8.104771e-08 5.636475e-05 -12.465510
## 5 6.717552e-08 5.636475e-05  -5.497636
## 6 4.567204e-08 5.636475e-05  -7.130614</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">with</span>(res, <span class="kw">plot</span>(lfc, <span class="op">-</span><span class="kw">log10</span>(adjp)))
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>))
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="op">-</span><span class="kw">log10</span>(<span class="fl">0.001</span>))</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-112"></span>
<p class="caption marginnote shownote">
Figure 12.8: Volcano plot illustrating how stimulation down-regulates the expression of numerous genes.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-112-1.png" alt="Volcano plot illustrating how stimulation down-regulates the expression of numerous genes." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Considering that a gene is called differentially expressed if it has an absolute log2 fold-change &gt; 1 and an adjusted p-value &lt; 0.001, how many differentially genes are there?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-84" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-84', 'sol-start-84')"></span>
</p>
<div id="sol-body-84" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(adjp <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.001</span>, <span class="kw">abs</span>(lfc) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>nrow</code></pre></div>
<pre><code>## [1] 133</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise the expression of these differentially expressed genes in all sampes using a heatmap and interpret the figure.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-85" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-85', 'sol-start-85')"></span>
</p>
<div id="sol-body-85" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">k_de &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">rownames_to_column</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(adjp <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.001</span>, <span class="kw">abs</span>(lfc) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(rowname) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">unlist</span>()

<span class="kw">heatmap</span>(e[k_de, ])</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-114"></span>
<p class="caption marginnote shownote">
Figure 12.9: Heatmap of the 73 differentially expressed genes.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-114-1.png" alt="Heatmap of the 73 differentially expressed genes." width="672">
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"ComplexHeatmap"</span>)
col_annot &lt;-<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">cell_type =</span> pd<span class="op">$</span>cell_type, <span class="dt">treatment =</span> pd<span class="op">$</span>treatment)
<span class="kw">Heatmap</span>(e[k_de, ], <span class="dt">top_annotation =</span> col_annot)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-115"></span>
<p class="caption marginnote shownote">
Figure 12.10: Annotated heatmap of the 73 differentially expressed genes.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-115-1.png" alt="Annotated heatmap of the 73 differentially expressed genes." width="672">
</div>
<p>We see that most genes that are differentially expressed in cell type A are show a consistent pattern in cell type B, except for the botton 5 genes, that show a low expression accross both conditions.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Select the 3 genes with the smallest p-values and visualise their expression in all samples (cell types A and B, both stimulated and un-stimulated) using boxplots.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-86" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-86', 'sol-start-86')"></span>
</p>
<div id="sol-body-86" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">k_<span class="dv">3</span> &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">rownames_to_column</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">arrange</span>(adjp) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">head</span>(<span class="dv">3</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(rowname) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">unlist</span>()

e[k_<span class="dv">3</span>, ] <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">rownames_to_column</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">gather</span>(<span class="dt">value =</span> expression, <span class="dt">key =</span> sample_id, <span class="op">-</span>rowname) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">full_join</span>(pd) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> treatment, <span class="dt">y =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.size =</span> <span class="dv">-1</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_jitter</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_grid</span>(rowname <span class="op">~</span><span class="st"> </span>cell_type)</code></pre></div>
<pre><code>## Joining, by = "sample_id"</code></pre>
<div class="figure">
<span id="fig:unnamed-chunk-116"></span>
<p class="caption marginnote shownote">
Figure 12.11: Expression of the three most differentially expressed genes.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-116-1.png" alt="Expression of the three most differentially expressed genes." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Perform a k-means clustering of all genes and samples using k = 5 and visualise these clusters on a PCA plot. Interprete the figure.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-87" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-87', 'sol-start-87')"></span>
</p>
<div id="sol-body-87" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
res<span class="op">$</span>cl &lt;-<span class="st"> </span><span class="kw">kmeans</span>(e, <span class="dt">centers =</span> <span class="dv">5</span>, <span class="dt">nstart =</span> <span class="dv">10</span>)<span class="op">$</span>cl
pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(e, <span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)
<span class="kw">fviz_pca_ind</span>(pca, <span class="dt">habillage =</span> res<span class="op">$</span>cl, <span class="dt">geom =</span> <span class="st">"point"</span>)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-117"></span>
<p class="caption marginnote shownote">
Figure 12.12: PCA of the <em>kem2</em> genes displaying the 5 arbitrary clusters.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-117-1.png" alt="PCA of the *kem2* genes displaying the 5 arbitrary clusters." width="672">
</div>
<p>The clusters represent the diversity of gene expression along PC1. Cluster 3, on the left of the figure, seems to be the most variable and different, while to others represent a more continuous gradient along of gene expression variation.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Repeat the volcano plot above, colouring the genes based on their respective clusters. Refine your interpretation of the gene-level PCA above.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-88" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-88', 'sol-start-88')"></span>
</p>
<div id="sol-body-88" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">with</span>(res, <span class="kw">plot</span>(lfc, <span class="op">-</span><span class="kw">log10</span>(adjp), <span class="dt">col =</span> res<span class="op">$</span>cl))
<span class="kw">legend</span>(<span class="st">"bottomleft"</span>, <span class="dt">legend =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">pch =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-118"></span>
<p class="caption marginnote shownote">
Figure 12.13: Volcano plot displaying the five clusters
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-118-1.png" alt="Volcano plot displaying the five clusters" width="672">
</div>
<p>We see that the most diverse cluster corresponds roughly to the genes that show significant down regulation upon stimulation.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-biovis.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-si.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2020-01-08
</p>
</div>
</div>



</body>
</html>
