<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 3 High-level data structures | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-08-23" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 3 High-level data structures | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.46.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.46.1/plotly-latest.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a id="active-page" href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a><ul class="toc-sections">
<li class="toc"><a href="#introduction-1"> Introduction</a></li>
<li class="toc"><a href="#why-objects"> Why objects?</a></li>
<li class="toc"><a href="#examples-in-base-r"> Examples in base R</a></li>
<li class="toc"><a href="#examples-from-the-bioconductor-project"> Examples from the Bioconductor project</a></li>
<li class="toc"><a href="#implementations"> Implementations</a></li>
</ul>
<a href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with <em><span>Biostrings</span></em></a>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centering, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical learning</a>
<a href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-ul.html"><span class="toc-section-number">8</span> Unsupervised learning</a>
<a href="sec-dimred.html"><span class="toc-section-number">9</span> Dimensionality reduction</a>
<a href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:obj" class="section level1">
<h1>
<span class="header-section-number">Chapter 3</span> High-level data structures</h1>
<p><strong>Learning objectives</strong></p>
<ul>
<li>Understand the difference between basic data structures and
high-level objects.</li>
<li>Understand the concept of high-level object.</li>
<li>Know how to learn about objects and access data from them.</li>
<li>Learn about import Bioconductor objects widely used in omics data
analysis.</li>
<li>Become familiar with how objects are implemented.</li>
</ul>
<div id="introduction-1" class="section level2">
<h2>
<span class="header-section-number">3.1</span> Introduction</h2>
<p>The term <em>object</em> stems from the widely used objet-oriented
programming (OOP) paradigm. OOP is however outside of scope of this
course (see part III of <span class="citation">(Wickham <label for="tufte-mn-12" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-12" class="margin-toggle">2014<span class="marginnote">Wickham, Hadley. 2014. <em>Advanced R</em>. Chapman &amp; Hall/Crc the R Series. Taylor &amp; Francis.</span>)</span>, which is also <a href="https://adv-r.hadley.nz/">freely available
online</a>). In the <em>R refresher</em> chapter
<a href="sec-refresher.html#sec:refresher">1</a>, we have seen the base types of R, which we have
efficiently used in many situations. Sometimes, however, these aren’t
enough.</p>
<p>We have also use object, for example in chapter 2 <a href="sec-vis.html#sec:vis">2</a>, when
generating ggplot visualisation. We have seen that it is possible to
store the output of <code>ggplot2</code> in a variable, that contain all the data
and information necessary to produce the figure.</p>
</div>
<div id="why-objects" class="section level2">
<h2>
<span class="header-section-number">3.2</span> Why objects?</h2>
<ul>
<li><p>Combine many parts of a complext data, for example store data (and
possibly different forms thereof (see chapter <a href="sec-norm.html#sec:norm">5</a>) and
metadata together.</p></li>
<li><p>Objects are <strong>self describing</strong>: an object to store high-throughput
sequencing (HTS) data contains high-throughput sequencing
data. Replace high-throughput sequencing data by DNA sequences (see
chapter <a href="sec-biostrings.html#sec:biostrings">4</a>), proteomics, … and it will still
hold.</p></li>
<li><p>Using object enables <strong>consistency</strong><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>: whoever prepares the
HTS data, it will always have the same structure.</p></li>
</ul>
<ul>
<li>If all information is stored in a single object<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a>, it is
easier to save it, to share it with others, and to use it as an
input to a function.</li>
</ul>
<p>The above properties lead to greater <strong>robustness</strong> and
<strong>interoperability</strong> when working with objects modelling complex data.</p>
<ul>
<li><p>Common behaviour (functions) can be specialised to match the
particular data type (this is called <strong>polymorphism</strong>): for example
plotting (a vector, a matrix, NGS data), summary, subsetting (<code>[</code>
and <code>[[</code>), …</p></li>
<li><p>Another import benefit of using objects to model complex data is
<strong>abstraction</strong>: it is possible to hide all the (unnecessary for its
usage) complexity of a particular type of data and present/give
access to its most imporant parts in a user-friendly <strong>interface</strong>
and efficient way.</p></li>
<li><p>Relations between object can be recorded (this is called
<strong>inheritance</strong>).</p></li>
</ul>
<p>To avoid confusion, it is useful to define and separate objects and
classes. A <strong>class</strong> describes (models) the general properties for
groups of objects. For example, we could define a class to describe
people in this course as follows:</p>
<pre><code>Class person:
- Name: a character
- Surname: a character
- Noma: an optional numeric
- Role: a character (student, teaching assistant, instructor)</code></pre>
<p>An <strong>object</strong> is and instantiation of a class with specific values in
its fields:</p>
<pre><code>Object of class person:
- Name: Laurent
- Surname: Gatto
- Noma: NA
- Role: instructor</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Describe two objects that would describe a teaching assistant and a
student.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-7" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-7', 'sol-start-7')"></span>
</p>
<div id="sol-body-7" class="solution-body" style="display: none;">
<pre><code>Object of class person:
- Name: Axelle
- Surname: Loriot
- Noma: NA
- Role: teaching assistant

Object of class person:
- Name: Object of class person:
- Name: Alejandra
- Surname: Sanchez-Cortez
- Noma: 123456
- Role: student</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>We will focus on objects and their usage here.</p>
</div>
<div id="examples-in-base-r" class="section level2">
<h2>
<span class="header-section-number">3.3</span> Examples in base R</h2>
<p>Let’s start by generating a testing dataset composed of two correlated
variables, <code>x</code> and <code>y</code>:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="sec-obj.html#cb50-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb50-2"><a href="sec-obj.html#cb50-2"></a>x &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="dv">100</span>)</span>
<span id="cb50-3"><a href="sec-obj.html#cb50-3"></a>y &lt;-<span class="st"> </span>x <span class="op">+</span><span class="st"> </span><span class="kw">rnorm</span>(<span class="dv">100</span>)</span>
<span id="cb50-4"><a href="sec-obj.html#cb50-4"></a>df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(x, y)</span>
<span id="cb50-5"><a href="sec-obj.html#cb50-5"></a><span class="kw">head</span>(df)</span></code></pre></div>
<pre><code>##            x          y
## 1 -0.6264538 -1.2468205
## 2  0.1836433  0.2257592
## 3 -0.8356286 -1.7465503
## 4  1.5952808  1.7533096
## 5  0.3295078 -0.3250769
## 6 -0.8204684  0.9468189</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="sec-obj.html#cb52-1"></a><span class="kw">plot</span>(y <span class="op">~</span><span class="st"> </span>x, df)</span></code></pre></div>
<p><img src="WSBIM1207_files/figure-html/df-1.png" width="672"></p>
<p>We are now going to model the linear relation between <code>x</code> and <code>y</code>
using the <code>lm</code> (linear model) function. The resulting variable <code>fit</code>
is the result of computing the linear regression
<span class="math inline">\(y = \beta_0 + beta_1 x + \epsilon\)</span>:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="sec-obj.html#cb53-1"></a>fit &lt;-<span class="st"> </span><span class="kw">lm</span>(y <span class="op">~</span><span class="st"> </span>x, df)</span></code></pre></div>
</div>
<div id="examples-from-the-bioconductor-project" class="section level2">
<h2>
<span class="header-section-number">3.4</span> Examples from the Bioconductor project</h2>
</div>
<div id="implementations" class="section level2">
<h2>
<span class="header-section-number">3.5</span> Implementations</h2>
<div id="s3-objects" class="section level3">
<h3>
<span class="header-section-number">3.5.1</span> S3 objects</h3>
</div>
<div id="s4-objects" class="section level3">
<h3>
<span class="header-section-number">3.5.2</span> S4 objects</h3>

</div>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-vis.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-biostrings.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-08-23
</p>
</div>
</div>



</body>
</html>
