<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 Manipulating sequences with Biostrings | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-09-30" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 4 Manipulating sequences with Biostrings | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a>
<a id="active-page" href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a><ul class="toc-sections">
<li class="toc"><a href="#working-with-sequences---a-first-start"> Working with sequences - a first start</a></li>
<li class="toc"><a href="#working-with-sequences---using-biostrings"> Working with sequences - using <code>Biostrings</code></a></li>
<li class="toc"><a href="#reading-dna-sequence-data-from-a-file"> Reading DNA sequence data from a file</a></li>
<li class="toc"><a href="#tidy-bioconductor"> Tidy <em>Bioconductor</em>?</a></li>
<li class="toc"><a href="#working-with-genomes"> Working with genomes</a></li>
<li class="toc"><a href="#additional-exercises-3"> Additional exercises</a></li>
</ul>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centering, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-ul.html"><span class="toc-section-number">8</span> Unsupervised learning</a>
<a href="sec-dimred.html"><span class="toc-section-number">9</span> Dimensionality reduction</a>
<a href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:biostrings" class="section level1">
<h1>
<span class="header-section-number">Chapter 4</span> Manipulating sequences with Biostrings</h1>
<p><strong>Learning objectives</strong></p>
<p>In this chapter, we will learn about a new object, that is
specialised for the manipulation of biological sequences.</p>
<ul>
<li>Learn about the fasta format format for biological sequences.</li>
<li>Learn about the <code>*StingSet</code> classes.</li>
<li>Working with whole genomes.</li>
<li>Find patterns in sequence data.</li>
</ul>
<p>This chapter is based on the course material presented by Martin
Morgan during the <a href="https://uclouvain-cbio.github.io/BSS2019/">Bioinformatics Summer School
2019</a> at Louvain-la-Neuve,
Belgium.</p>
<div id="working-with-sequences---a-first-start" class="section level2">
<h2>
<span class="header-section-number">4.1</span> Working with sequences - a first start</h2>
<p>It is easy to use standard R character vectors to define DNA
sequences. Below, we create a vector with three short sequences.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="sec-biostrings.html#cb120-1"></a>sequences &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"AAATCGA"</span>, <span class="st">"ATACAACAT"</span>, <span class="st">"TTGCCA"</span>)</span>
<span id="cb120-2"><a href="sec-biostrings.html#cb120-2"></a>sequences</span></code></pre></div>
<pre><code>## [1] "AAATCGA"   "ATACAACAT" "TTGCCA"</code></pre>
<p>We can ask about properties of these sequences and perform some
operations using base R function:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="sec-biostrings.html#cb122-1"></a><span class="kw">length</span>(sequences)</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="sec-biostrings.html#cb124-1"></a><span class="kw">nchar</span>(sequences)</span></code></pre></div>
<pre><code>## [1] 7 9 6</code></pre>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="sec-biostrings.html#cb126-1"></a>sequences[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)]</span></code></pre></div>
<pre><code>## [1] "AAATCGA" "TTGCCA"</code></pre>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="sec-biostrings.html#cb128-1"></a><span class="kw">sample</span>(sequences)</span></code></pre></div>
<pre><code>## [1] "TTGCCA"    "AAATCGA"   "ATACAACAT"</code></pre>
<p>However, once is we want to perform biologically relevant operations,
base R fails us; it has no notion of operations relevant to DNA
sequences, e.g.,</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="sec-biostrings.html#cb130-1"></a><span class="kw">reverseComplement</span>(sequences)</span></code></pre></div>
<p>Likewise, we can name a variable anything, the semantic meaning
of the variable name is not enforced by R</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="sec-biostrings.html#cb131-1"></a>not_a_dna_sequence &lt;-<span class="st"> "I love R"</span></span></code></pre></div>
</div>
<div id="working-with-sequences---using-biostrings" class="section level2">
<h2>
<span class="header-section-number">4.2</span> Working with sequences - using <code>Biostrings</code>
</h2>
<p>The Bioconductor <em><a href="https://bioconductor.org/packages/3.10/Biostrings">Biostrings</a></em>
package can do all that base R can do, in addition to knowing about
the semantics of the sequences is handles. Let’s start by loading the
package:</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="sec-biostrings.html#cb132-1"></a><span class="kw">library</span>(<span class="st">"Biostrings"</span>)</span></code></pre></div>
<div id="dnastringset" class="section level3">
<h3>
<span class="header-section-number">4.2.1</span> <code>DNAStringSet()</code>
</h3>
<p>The <code>DNAStringSet</code> class that <em>knows</em> about DNA sequences. We can
easily create a <code>DNAStringSet</code> from our character vector with</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="sec-biostrings.html#cb133-1"></a>dna &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(sequences)</span>
<span id="cb133-2"><a href="sec-biostrings.html#cb133-2"></a>dna</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 3
##     width seq
## [1]     7 AAATCGA
## [2]     9 ATACAACAT
## [3]     6 TTGCCA</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Does the object <code>dna</code> support the operations illustrated above for a
character vector, especially <code>length()</code>, <code>nchar()</code>, <code>[</code>, and
<code>sample()</code>?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-14" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-14', 'sol-start-14')"></span>
</p>
<div id="sol-body-14" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="sec-biostrings.html#cb135-1"></a><span class="kw">length</span>(dna)</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="sec-biostrings.html#cb137-1"></a><span class="kw">nchar</span>(dna)</span></code></pre></div>
<pre><code>## [1] 7 9 6</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="sec-biostrings.html#cb139-1"></a>dna[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)]</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]     7 AAATCGA
## [2]     6 TTGCCA</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="sec-biostrings.html#cb141-1"></a><span class="kw">sample</span>(dna)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 3
##     width seq
## [1]     9 ATACAACAT
## [2]     6 TTGCCA
## [3]     7 AAATCGA</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Prove to yourself that at least some other useful, DNA-specific,
functions exist, e.g., <code>reverse()</code> and <code>reverseComplement()</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-15" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-15', 'sol-start-15')"></span>
</p>
<div id="sol-body-15" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="sec-biostrings.html#cb143-1"></a><span class="kw">reverse</span>(dna)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 3
##     width seq
## [1]     7 AGCTAAA
## [2]     9 TACAACATA
## [3]     6 ACCGTT</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="sec-biostrings.html#cb145-1"></a><span class="kw">reverseComplement</span>(dna)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 3
##     width seq
## [1]     7 TCGATTT
## [2]     9 ATGTTGTAT
## [3]     6 TGGCAA</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What happens when you try to create a <code>DNAStringSet()</code> from an object
such as <code>not_a_dna_sequence</code>, defined above, that does not contain a
DNA sequence? Warning: the message is quite cryptic, can you provide a
‘human’ translation?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-16" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-16', 'sol-start-16')"></span>
</p>
<div id="sol-body-16" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="sec-biostrings.html#cb147-1"></a><span class="kw">try</span>(<span class="kw">DNAStringSet</span>(not_a_dna_sequence))</span></code></pre></div>
<pre><code>## Error in .Call2("new_XString_from_CHARACTER", class(x0), string, start,  : 
##   key 73 (char 'I') not in lookup table</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Why does <code>DNAStringSet("ACGTMRW")</code> not create an error, since <code>MRW</code>
are not standard nucleotides? For hints, see the section ’The DNA
alphabet:" on the help page <code>?DNAString</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What is the difference between a <code>DNAString</code> and a <code>DNAStringSet</code>?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-17" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-17', 'sol-start-17')"></span>
</p>
<div id="sol-body-17" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="sec-biostrings.html#cb149-1"></a><span class="kw">DNAString</span>(<span class="st">"AGACCATG"</span>)</span></code></pre></div>
<pre><code>##   8-letter "DNAString" instance
## seq: AGACCATG</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="sec-biostrings.html#cb151-1"></a><span class="kw">DNAStringSet</span>(<span class="kw">c</span>(<span class="st">"AGACCATG"</span>, <span class="st">"AAAAAATTTA"</span>))</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]     8 AGACCATG
## [2]    10 AAAAAATTTA</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="learning-more-about-a-class" class="section level3">
<h3>
<span class="header-section-number">4.2.2</span> Learning more about a class</h3>
<p>The function <code>DNAStringSet()</code> returns an <em>object</em> that has a
particular class</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="sec-biostrings.html#cb153-1"></a><span class="kw">class</span>(dna)</span></code></pre></div>
<pre><code>## [1] "DNAStringSet"
## attr(,"package")
## [1] "Biostrings"</code></pre>
<p>Associated with the class are a series of <em>methods</em> (i.e. functions
that have a behaviour that is tuned for that class) that operate on
the class.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Discover many (unfortunately, not all) methods acting on
an object of class <code>DNAStringSet</code> using <code>methods(class = "DNAStringSet")</code>. Verify that <code>reverseComplement</code> is among those
methods.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Help pages describing a particular method can be found using <code>?</code>, with
the search query quoted and with tab-completion providing hints on
what the appropriate help topic is.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Find the help page for the <code>reverseComplement</code> method operating on a
<code>DNAStringSet</code> object, using <code>?reverseComplement</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Help pages provide a description of the technical details required for
creating classes and using methods. Vignettes provide a more narrative
description of overall package use.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use <code>browseVignettes(package = "Biostrings")</code> to see vignettes
available for this package; explore a few vignettes to get a sense of
possible content.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
</div>
<div id="reading-dna-sequence-data-from-a-file" class="section level2">
<h2>
<span class="header-section-number">4.3</span> Reading DNA sequence data from a file</h2>
<p>It is unlikely that we would enter 1000’s of DNA sequences ‘by
hand’. Instead, we might read the data from a standard file
format. For DNA sequences the standard file format is often a ‘FASTA’
file, sometimes abbreviated with an extension <code>.fa</code> and often
compressed with an additional extension <code>.fa.gz</code>. An example of a
FASTA file containing DNA sequences of the 2000bp upstream nucleotides
of all genes annotated in the <em>Drosophila melanogaster</em> <code>dm3</code> genome
build, is distributed with the <code>Biostrings</code> package. Here’s the path
to the FASTA file.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="sec-biostrings.html#cb155-1"></a>fa_file &lt;-</span>
<span id="cb155-2"><a href="sec-biostrings.html#cb155-2"></a><span class="st">    </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">"Biostrings"</span>, <span class="st">"extdata"</span>, <span class="st">"dm3_upstream2000.fa.gz"</span>)</span></code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Take a peak at the structure of a FASTA file by looking at the first
five lines. You can use the <code>readLines</code> function, setting the <code>n</code>
parameter to read only a limited number of files (see <code>readLines</code>).</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-18" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-18', 'sol-start-18')"></span>
</p>
<div id="sol-body-18" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="sec-biostrings.html#cb156-1"></a><span class="kw">readLines</span>(fa_file, <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## [1] "&gt;NM_078863_up_2000_chr2L_16764737_f chr2L:16764737-16766736"
## [2] "gttggtggcccaccagtgccaaaatacacaagaagaagaaacagcatctt"         
## [3] "gacactaaaatgcaaaaattgctttgcgtcaatgactcaaaacgaaaatg"         
## [4] "tttttgtgcttttcgaacaaaaaattgggagcagaatattggattcgctt"         
## [5] "ttttcgcatcgaatatcttaagggagcaagggaagaaccatctagaataa"</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>In this case, the first line is an identifier, containing information
about the gene <code>NM_078863</code> as well as the genomic coordinates of the
sequence <code>chr2L:16764737-16766736</code>. The next lines are the DNA
sequence. After a certain number of lines, a new record starts.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="sec-biostrings.html#cb158-1"></a><span class="kw">tail</span>(<span class="kw">readLines</span>(fa_file, <span class="dv">44</span>), <span class="dv">5</span>)</span></code></pre></div>
<pre><code>## [1] "cacgcacaccgatcgtcgaatcgaaaagctttcggggtcttacgggatcc"         
## [2] "atgggtatcaagttgccccgtataaaaggcaagtttaccggttgcacggt"         
## [3] "&gt;NM_001201794_up_2000_chr2L_8382455_f chr2L:8382455-8384454"
## [4] "ttatttatgtaggcgcccgttcccgcagccaaagcactcagaattccggg"         
## [5] "cgtgtagcgcaacgaccatctacaaggcaatattttgatcgcttgttagg"</code></pre>
<p>We could fairly <em>easily</em> write our own parser for this format, but
this would be error-prone and unnecessary. Instread, we want to use
the <code>readDNAStringSet</code> function from the <code>Biostrings</code> package for
that.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="sec-biostrings.html#cb160-1"></a>dna &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fa_file)</span>
<span id="cb160-2"><a href="sec-biostrings.html#cb160-2"></a>dna</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 26454
##         width seq                                      names               
##     [1]  2000 GTTGGTGGCCCACCAGTGC...GTTTACCGGTTGCACGGT NM_078863_up_2000...
##     [2]  2000 TTATTTATGTAGGCGCCCG...CGGAAAGTCATCCTCGAT NM_001201794_up_2...
##     [3]  2000 TTATTTATGTAGGCGCCCG...CGGAAAGTCATCCTCGAT NM_001201795_up_2...
##     [4]  2000 TTATTTATGTAGGCGCCCG...CGGAAAGTCATCCTCGAT NM_001201796_up_2...
##     [5]  2000 TTATTTATGTAGGCGCCCG...CGGAAAGTCATCCTCGAT NM_001201797_up_2...
##     ...   ... ...
## [26450]  2000 ATTTACAAGACTAATAAAG...ATTAAATTTCAATAAAAC NM_001111010_up_2...
## [26451]  2000 GATATACGAAGGACGACCT...TTTGAGTTGTTATATATT NM_001015258_up_2...
## [26452]  2000 GATATACGAAGGACGACCT...TTTGAGTTGTTATATATT NM_001110997_up_2...
## [26453]  2000 GATATACGAAGGACGACCT...TTTGAGTTGTTATATATT NM_001276245_up_2...
## [26454]  2000 CGTATGTATTAGTTAACTC...AAGTGTAAGAACAAATTG NM_001015497_up_2...</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Query the object for basic properties, e.g., it’s <code>length()</code> and that
the number of characters in each sequence is 2000
<code>unique(nchar(dna))</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use <code>letterFrequency()</code> to determine GC content of each of the DNA
sequences in <code>dna</code>. The <code>letters</code> argument should be <code>"GC"</code>; <code>as.prob = TRUE</code> returns values between 0 and 1. The data is returned as a
matrix with 1 column.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-19" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-19', 'sol-start-19')"></span>
</p>
<div id="sol-body-19" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="sec-biostrings.html#cb162-1"></a>gc_content &lt;-<span class="st"> </span><span class="kw">letterFrequency</span>(dna, <span class="dt">letters =</span> <span class="st">"GC"</span>, <span class="dt">as.prob =</span> <span class="ot">TRUE</span>)</span>
<span id="cb162-2"><a href="sec-biostrings.html#cb162-2"></a>gc_content</span></code></pre></div>
<pre><code>##                G|C
##     [1,] 0.3785000
##     [2,] 0.4300000
##     [3,] 0.4300000
##     [4,] 0.4300000
##     [5,] 0.4300000
##     [6,] 0.4300000
##     [7,] 0.4300000
##     [8,] 0.4300000
##     [9,] 0.4300000
##    [10,] 0.4300000
##    [11,] 0.3645000
##    [12,] 0.3645000
##    [13,] 0.3785000
##    [14,] 0.3785000
##    [15,] 0.3785000
##    [16,] 0.3785000
##    [17,] 0.3785000
##    [18,] 0.3930000
##    [19,] 0.3785000
##    [20,] 0.3785000
##    [21,] 0.3785000
##    [22,] 0.3775000
##    [23,] 0.3775000
##    [24,] 0.3775000
##    [25,] 0.3785000
##    [26,] 0.3785000
##    [27,] 0.3790000
##    [28,] 0.3625000
##    [29,] 0.3930000
##    [30,] 0.3930000
##    [31,] 0.3785000
##    [32,] 0.4010000
##    [33,] 0.4010000
##    [34,] 0.3550000
##    [35,] 0.3550000
##    [36,] 0.3790000
##    [37,] 0.3550000
##    [38,] 0.3790000
##    [39,] 0.4230000
##    [40,] 0.3820000
##    [41,] 0.3875000
##    [42,] 0.3875000
##    [43,] 0.3875000
##    [44,] 0.4440000
##    [45,] 0.4440000
##    [46,] 0.4440000
##    [47,] 0.4440000
##    [48,] 0.3875000
##    [49,] 0.3665000
##    [50,] 0.4530000
##    [51,] 0.3240000
##    [52,] 0.4240000
##    [53,] 0.4240000
##    [54,] 0.3360000
##    [55,] 0.4035000
##    [56,] 0.4660000
##    [57,] 0.3995000
##    [58,] 0.3995000
##    [59,] 0.3950000
##    [60,] 0.3995000
##    [61,] 0.3835000
##    [62,] 0.3240000
##    [63,] 0.3360000
##    [64,] 0.3360000
##    [65,] 0.3950000
##    [66,] 0.3360000
##    [67,] 0.4240000
##    [68,] 0.3985000
##    [69,] 0.5060000
##    [70,] 0.5060000
##    [71,] 0.4505000
##    [72,] 0.4105000
##    [73,] 0.3980000
##    [74,] 0.3985000
##    [75,] 0.3985000
##    [76,] 0.3965000
##    [77,] 0.4070000
##    [78,] 0.4070000
##    [79,] 0.4245000
##    [80,] 0.3680000
##    [81,] 0.3595000
##    [82,] 0.3850000
##    [83,] 0.3915000
##    [84,] 0.3965000
##    [85,] 0.3900000
##    [86,] 0.3920000
##    [87,] 0.3920000
##    [88,] 0.4805000
##    [89,] 0.4805000
##    [90,] 0.4220000
##    [91,] 0.4340000
##    [92,] 0.4340000
##    [93,] 0.4290000
##    [94,] 0.4190000
##    [95,] 0.4520000
##    [96,] 0.3425000
##    [97,] 0.4450000
##    [98,] 0.4600000
##    [99,] 0.4560000
##   [100,] 0.3920000
##  [ reached getOption("max.print") -- omitted 26354 rows ]</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>What are the mean and standard deviation of the GC contents in the 26454 genes?
What genes(s) has the highes GC content?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-20" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-20', 'sol-start-20')"></span>
</p>
<div id="sol-body-20" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="sec-biostrings.html#cb164-1"></a><span class="kw">mean</span>(gc_content)</span></code></pre></div>
<pre><code>## [1] 0.4228301</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="sec-biostrings.html#cb166-1"></a><span class="kw">sd</span>(gc_content)</span></code></pre></div>
<pre><code>## [1] 0.04935743</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="sec-biostrings.html#cb168-1"></a><span class="kw">range</span>(gc_content)</span></code></pre></div>
<pre><code>## [1] 0.0800 0.6145</code></pre>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="sec-biostrings.html#cb170-1"></a>dna[<span class="kw">which</span>(gc_content <span class="op">==</span><span class="st"> </span><span class="kw">max</span>(gc_content))]</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 1
##     width seq                                          names               
## [1]  2000 CGCTACTACGGCGCCAGTTCG...CTTGCTAAAGCACTCTGGTC NM_133087_up_2000...</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Plot the distribution of GC frequencies in the <code>dna</code> object using base
graphics <code>hist()</code> and <code>plot(density())</code>, and using <code>ggplot()</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-21" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-21', 'sol-start-21')"></span>
</p>
<div id="sol-body-21" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="sec-biostrings.html#cb172-1"></a><span class="kw">hist</span>(gc_content)</span></code></pre></div>
<p><img src="WSBIM1322_files/figure-html/letterfreqex3a-1.png" width="672"></p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="sec-biostrings.html#cb173-1"></a><span class="kw">plot</span>(<span class="kw">density</span>(gc_content))</span></code></pre></div>
<p><img src="WSBIM1322_files/figure-html/letterfreqex3a-2.png" width="672"></p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="sec-biostrings.html#cb174-1"></a><span class="kw">library</span>(<span class="st">"tidyverse"</span>) </span>
<span id="cb174-2"><a href="sec-biostrings.html#cb174-2"></a>p &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(gc_content) <span class="op">%&gt;%</span></span>
<span id="cb174-3"><a href="sec-biostrings.html#cb174-3"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="st">`</span><span class="dt">G|C</span><span class="st">`</span>))</span>
<span id="cb174-4"><a href="sec-biostrings.html#cb174-4"></a>p <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>()</span></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="WSBIM1322_files/figure-html/letterfreqex3b-1.png" width="672"></p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="sec-biostrings.html#cb176-1"></a>p <span class="op">+</span><span class="st"> </span><span class="kw">geom_density</span>()</span></code></pre></div>
<p><img src="WSBIM1322_files/figure-html/letterfreqex3b-2.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="tidy-bioconductor" class="section level2">
<h2>
<span class="header-section-number">4.4</span> Tidy <em>Bioconductor</em>?</h2>
<p>Although <em>Bioconductor</em> emphasizes formal objects like <code>DNAStringSet</code>
rather than <code>tibble</code>-like data frames, some of the ways one interacts
with tidy data can be applied to <em>Bioconductor</em> objects. For instance,
the GC content example might be written in ‘traditional’ form as</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="sec-biostrings.html#cb177-1"></a>gc &lt;-<span class="st"> </span><span class="kw">letterFrequency</span>(dna, <span class="st">"GC"</span>, <span class="dt">as.prob =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>but could be written using pipes and to reesult in a tibble for easier
down-stream manipulation</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="sec-biostrings.html#cb178-1"></a>gc &lt;-<span class="st"> </span></span>
<span id="cb178-2"><a href="sec-biostrings.html#cb178-2"></a><span class="st">    </span>dna <span class="op">%&gt;%</span></span>
<span id="cb178-3"><a href="sec-biostrings.html#cb178-3"></a><span class="st">    </span><span class="kw">letterFrequency</span>(<span class="st">"GC"</span>, <span class="dt">as.prob =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></span>
<span id="cb178-4"><a href="sec-biostrings.html#cb178-4"></a><span class="st">    </span>tibble<span class="op">::</span><span class="kw">as_tibble</span>()</span>
<span id="cb178-5"><a href="sec-biostrings.html#cb178-5"></a>gc</span></code></pre></div>
<pre><code>## # A tibble: 26,454 x 1
##    `G|C`
##    &lt;dbl&gt;
##  1 0.378
##  2 0.43 
##  3 0.43 
##  4 0.43 
##  5 0.43 
##  6 0.43 
##  7 0.43 
##  8 0.43 
##  9 0.43 
## 10 0.43 
## # … with 26,444 more rows</code></pre>
</div>
<div id="working-with-genomes" class="section level2">
<h2>
<span class="header-section-number">4.5</span> Working with genomes</h2>
<p><code>BSgenome</code> packages inside Bioconductor contain whole genome sequences
as distributed by ENSEMBL, NCBI and others. In this next example we
will load the whole genome sequence for Homo sapiens from UCSC’s hg19
build, and calculate the GC content across chromosome 14.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="sec-biostrings.html#cb180-1"></a><span class="kw">library</span>(<span class="st">"Biostrings"</span>)</span>
<span id="cb180-2"><a href="sec-biostrings.html#cb180-2"></a><span class="kw">library</span>(<span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)</span>
<span id="cb180-3"><a href="sec-biostrings.html#cb180-3"></a>chr14_range &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="st">"chr14"</span>, <span class="kw">IRanges</span>(<span class="dv">1</span>, <span class="kw">seqlengths</span>(Hsapiens)[<span class="st">"chr14"</span>]))</span>
<span id="cb180-4"><a href="sec-biostrings.html#cb180-4"></a>chr14_range</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames      ranges strand
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]    chr14 1-107349540      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="sec-biostrings.html#cb182-1"></a>chr14_dna &lt;-<span class="st"> </span><span class="kw">getSeq</span>(Hsapiens, chr14_range)</span>
<span id="cb182-2"><a href="sec-biostrings.html#cb182-2"></a>chr14_dna</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 1
##         width seq
## [1] 107349540 NNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNN</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="sec-biostrings.html#cb184-1"></a><span class="kw">letterFrequency</span>(chr14_dna, <span class="st">"GC"</span>, <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>##           G|C
## [1,] 0.336276</code></pre>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Using <code>letterFrequency</code>, calculate the number and frequency of A, G, C
and T nucleotides in chromosome 14.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-22" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-22', 'sol-start-22')"></span>
</p>
<div id="sol-body-22" class="solution-body" style="display: none;">
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="sec-biostrings.html#cb186-1"></a><span class="kw">letterFrequency</span>(chr14_dna, <span class="kw">c</span>(<span class="st">"A"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>, <span class="st">"T"</span>))</span></code></pre></div>
<pre><code>##             A        G        C        T
## [1,] 25992966 18071947 18027132 26197495</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="sec-biostrings.html#cb188-1"></a><span class="kw">letterFrequency</span>(chr14_dna, <span class="kw">c</span>(<span class="st">"A"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>, <span class="st">"T"</span>), <span class="dt">as.prob =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>##             A        G        C        T
## [1,] 25992966 18071947 18027132 26197495</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="additional-exercises-3" class="section level2">
<h2>
<span class="header-section-number">4.6</span> Additional exercises</h2>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visit the UniProt web page at
<a href="https://www.uniprot.org/">https://www.uniprot.org/</a> and download the
manually reviewed complete human proteome (containing about 20500
proteins). Do to so, choose <em>Proteomes</em> on the main page, then search
for <em>Homo sapiens</em>, select it among the list of hits, choose the
reviewed Swiss-Prot release, and download all entries in compressed
fasta format.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the file into R. Beware though that these are protein sequences,
composed of amino acids. Read the <code>?readDNAStringSet</code> manual page to
find a more appropriate function.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-obj.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-norm.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-09-30
</p>
</div>
</div>



</body>
</html>
