<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 10 Supervised learning | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-11-28" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 10 Supervised learning | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a>
<a href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-dimred.html"><span class="toc-section-number">8</span> Unsupervised learning: dimensionality reduction</a>
<a href="sec-ul.html"><span class="toc-section-number">9</span> Unsupervised learning: clustering</a>
<a id="active-page" href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a><ul class="toc-sections">
<li class="toc"><a href="#introduction-5"> Introduction</a></li>
<li class="toc"><a href="#k-nearest-neighbours-classifier"> K-nearest neighbours classifier</a></li>
</ul>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:sl" class="section level1">
<h1>
<span class="header-section-number">Chapter 10</span> Supervised learning</h1>
<p>The goal of this last chapter about machine learning will be supervised learning, and in particular <strong>classification</strong>. In this chapter, you will learn</p>
<ul>
<li>learn what classification is;</li>
<li>learn about labelled, unlabelled, training and testing data;</li>
<li>learn about and apply k-nearest neighbour, a simple non-parametric classifier;</li>
<li>see why and how to use cross-validation;</li>
<li>and learn about model complexity and generalisation.</li>
</ul>
<div id="introduction-5" class="section level2">
<h2>
<span class="header-section-number">10.1</span> Introduction</h2>
<p>Often, we faced with omics data, we have annotations for some of the samples (or features) we have acquired:</p>
<ul>
<li><p>Among the 100s of patient tumours that were assayed using RNA sequencing of microarrays, we know the grade of the tumour for about half. We want to predict the grade of the other half using the gene expression profiles.</p></li>
<li><p>We have performed a spatial proteomics experiment such as in Christoforou <em>et al.</em> <span class="citation">(Christoforou et al. <label for="tufte-mn-28" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-28" class="margin-toggle">2016<span class="marginnote">Christoforou, A, C M Mulvey, L M Breckels, A Geladaki, T Hurrell, P C Hayward, T Naake, et al. 2016. “A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.” <em>Nat Commun</em> 7: 8992. doi:<a href="https://doi.org/10.1038/ncomms9992">10.1038/ncomms9992</a>.</span>)</span> (see section <a href="sec-dimred.html#sec:dimred02">8.7.2</a>) and know the sub-cellular localisation of some proteins. We want to predict the sub-cellular localisation of the other proteins using the protein profiles.</p></li>
</ul>
<p>In both of these examples, the quantitative data are the data that we want to use to predict properties about samples or features; these are called the <strong>predictors</strong>. The grade of the samples in the first example and the protein sub-cellular localisation in the second one are the <strong>labels</strong> that we have in some cases, and want to predcit otherwise. We can thus split our data in two parts, depending whether we have labels, or whether we want to predict them. The former are called <strong>labelled</strong>, and the latter <strong>unlabelled</strong>.</p>
<div class="figure">
<span id="fig:unnamed-chunk-75"></span>
<p class="caption marginnote shownote">
Figure 10.1: In supervised learning, the data are split in labelled or unlabelled data. The same applies when some of the columns are labelled.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-75-1.png" alt="In supervised learning, the data are split in labelled or unlabelled data. The same applies when some of the columns are labelled." width="1152">
</div>
<p>In the figure above, the labels represent categories that need to be inferred from the predictors. This class of supervised learning is called <strong>classification</strong>. Classification are also split into <strong>binary classification</strong> when there are only two classes, or <strong>multi-label</strong> problem when, like above, there are more than two. The latter is a generalisation of the binary task. When the annotations are continuous values, the situation is referred to as a <strong>regression</strong> problem.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>giris2</code> data from the <code>rWSBIM1322</code> package (requires version &gt;= 0.1.13). Identify the labelled and unlabelled data; how many are there respectively.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-65" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-65', 'sol-start-65')"></span>
</p>
<div id="sol-body-65" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"rWSBIM1322"</span>)
<span class="kw">data</span>(giris2)
<span class="kw">summary</span>(giris2)</code></pre></div>
<pre><code>##      BRCA1           BRCA2            TP53             A1CF         GRADE    
##  Min.   :3.870   Min.   :1.888   Min.   :0.4919   Min.   :0.1000   A   : 50  
##  1st Qu.:5.115   1st Qu.:2.700   1st Qu.:1.6242   1st Qu.:0.3044   B   : 50  
##  Median :5.800   Median :3.000   Median :4.3276   Median :1.3327   C   : 50  
##  Mean   :5.853   Mean   :3.058   Mean   :3.7796   Mean   :1.2111   NA's:100  
##  3rd Qu.:6.400   3rd Qu.:3.400   3rd Qu.:5.1885   3rd Qu.:1.8000             
##  Max.   :8.116   Max.   :4.400   Max.   :7.1766   Max.   :2.5685</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">is.na</span>(giris2<span class="op">$</span>GRADE))</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   150   100</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise the <code>giris2</code> data on a PCA plot, highlighting the labels (or absence thereof). Using the visualisation, will the classifying of the unlabelled data will be difficult or easy? Motivate your aswer.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-66" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-66', 'sol-start-66')"></span>
</p>
<div id="sol-body-66" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca2 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(giris2[, <span class="dv">-5</span>], <span class="dt">scale =</span> <span class="ot">TRUE</span>)
factoextra<span class="op">::</span><span class="kw">fviz_pca_ind</span>(pca2, <span class="dt">col.ind =</span> giris2<span class="op">$</span>GRADE)</code></pre></div>
<pre><code>## Warning: Removed 100 rows containing missing values (geom_point).</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_point).</code></pre>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-77-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="k-nearest-neighbours-classifier" class="section level2">
<h2>
<span class="header-section-number">10.2</span> K-nearest neighbours classifier</h2>
<p>In this chapter, we’ll use a simple, but useful classification algorithm, k-nearest neighbours (kNN) to classify the <em>giris</em> patients. We will use the <code>knn</code> function from the <em><a href="https://CRAN.R-project.org/package=class">class</a></em> package.</p>
<p>K-nearest neighbours works by directly measuring the (Euclidean) distance between observations and inferring the class of unlabelled data from the class of its nearest neighbours. In the figure below, the unlabelled instances <em>1</em> and <em>2</em> will be assigned classes <em>A</em> (blue) and <em>B</em> (red) as their closest neighbours are red and blue, respectively.</p>
<div class="figure">
<span id="fig:knnex"></span>
<p class="caption marginnote shownote">
Figure 10.2: Schematic illustrating the k nearest neighbors algorithm.
</p>
<img src="WSBIM1322_files/figure-html/knnex-1.png" alt="Schematic illustrating the k nearest neighbors algorithm." width="672">
</div>
<p>Typically in machine learning, there are two clear steps, where one first <strong>trains</strong> a model and then uses the model to <strong>predict</strong> new outputs (class labels in this case). In kNN, these two steps are combined into a single function call to <code>knn</code>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Separate the <code>giris2</code> data into two new datasets, one containing the labelled data that will be used to train the model and named <code>giris2_train</code>, and a second one containing the unlabelled data called <code>giris_test</code>.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-67" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-67', 'sol-start-67')"></span>
</p>
<div id="sol-body-67" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">giris2_labelled &lt;-<span class="st"> </span>giris2[<span class="op">!</span><span class="kw">is.na</span>(giris2<span class="op">$</span>GRADE), ]
giris2_unlabelled &lt;-<span class="st"> </span>giris2[<span class="kw">is.na</span>(giris2<span class="op">$</span>GRADE), ]</code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The <code>knn</code> function takes, among others<label for="tufte-sn-14" class="margin-toggle sidenote-number">14</label><input type="checkbox" id="tufte-sn-14" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">14</span> we will see additional ones later.</span>, the following arguments:</p>
<ol style="list-style-type: decimal">
<li>the labelled predictors, that will be used to <em>train</em> the model,</li>
<li>the unlabelled preductors, on which the model will be applied (see below why this is called <em>test</em>),</li>
<li>the labels (the length of this vector must match the number of rows of the labelled predictors).</li>
</ol>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Apply the kNN classifier on the <code>giris2</code> prepared in the previous exercise.</p></li>
<li><p>What is the class of the output?</p></li>
<li><p>How many of the unlabelled data have been assigned to the different grades?</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-68" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-68', 'sol-start-68')"></span>
</p>
<div id="sol-body-68" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"class"</span>)
knn_res &lt;-<span class="st"> </span><span class="kw">knn</span>(giris2_labelled[, <span class="dv">-5</span>],
               giris2_unlabelled[, <span class="dv">-5</span>],
               giris2_labelled[, <span class="dv">5</span>])
<span class="kw">class</span>(knn_res)</code></pre></div>
<pre><code>## [1] "factor"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(knn_res)</code></pre></div>
<pre><code>## knn_res
##  A  B  C 
## 31 38 31</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Visualise the results of the kNN classification on a PCA plot and interprete the results based on the firt PCA.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-69" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-69', 'sol-start-69')"></span>
</p>
<div id="sol-body-69" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">giris2_res &lt;-<span class="st"> </span>giris2
giris2_res[<span class="kw">is.na</span>(giris2<span class="op">$</span>GRADE), <span class="dv">5</span>] &lt;-<span class="st"> </span>knn_res
pca_knn &lt;-<span class="st"> </span><span class="kw">prcomp</span>(giris2_res[, <span class="dv">-5</span>], <span class="dt">scale =</span> <span class="ot">TRUE</span>)
p1 &lt;-<span class="st"> </span>factoextra<span class="op">::</span><span class="kw">fviz_pca_ind</span>(pca2, <span class="dt">col.ind =</span> giris2<span class="op">$</span>GRADE)
p2 &lt;-<span class="st"> </span>factoextra<span class="op">::</span><span class="kw">fviz_pca_ind</span>(pca_knn, <span class="dt">col.ind =</span> giris2_res<span class="op">$</span>GRADE)
<span class="kw">library</span>(<span class="st">"patchwork"</span>)
p1 <span class="op">/</span><span class="st"> </span>p2</code></pre></div>
<pre><code>## Warning: Removed 100 rows containing missing values (geom_point).</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_point).</code></pre>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-80-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question" id="">
<p class="question-begin">
► Question
</p>
<div class="question-body">


<ul>
<li><p>Can you identify some questionable results? Explore the results for patient 167, that was assigned group B.</p></li>
<li><p>Do do so, calculated the distances between sample 167 and all other labelled data.</p></li>
<li><p>Compare the labels of its 15 closest labelled data points.</p></li>
</ul>
<div class="question" id="">
<p class="question-begin">
► Question
</p>
<div class="question-body">






<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-70" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-70', 'sol-start-70')"></span>
</p>
<div id="sol-body-70" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## all pairwise distances
dd &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">dist</span>(giris2[, <span class="dv">-5</span>]))
## extract only those for sample 167
giris2<span class="op">$</span>dist_<span class="dv">167</span> &lt;-<span class="st"> </span>dd[<span class="dv">167</span>, ]

## look at the 15 nearest neighbours 
<span class="kw">library</span>(<span class="st">"tidyverse"</span>)
vote &lt;-<span class="st"> </span>giris2 <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">head</span>(<span class="dv">150</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">arrange</span>(dist_<span class="dv">167</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">head</span>(<span class="dv">15</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">vote =</span> <span class="kw">cumsum</span>(<span class="dt">count =</span> <span class="kw">ifelse</span>(GRADE <span class="op">==</span><span class="st"> "C"</span>, <span class="dv">1</span>, <span class="dv">-1</span>))) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">grade =</span> <span class="kw">ifelse</span>(vote <span class="op">&lt;=</span><span class="st"> </span><span class="dv">0</span>, <span class="st">"B"</span>, <span class="st">"C"</span>))
vote</code></pre></div>
<pre><code>##    BRCA1 BRCA2 TP53 A1CF GRADE  dist_167 vote grade
## 1    6.8   2.8  4.8  1.4     B 0.7239656   -1     B
## 2    6.7   3.0  5.0  1.7     B 0.7922045   -2     B
## 3    6.9   3.1  4.9  1.5     B 0.8065525   -3     B
## 4    6.7   2.5  5.8  1.8     C 0.8444024   -2     B
## 5    7.2   3.0  5.8  1.6     C 0.8592955   -1     B
## 6    6.9   3.1  5.4  2.1     C 0.8892644    0     B
## 7    6.8   3.0  5.5  2.1     C 0.8956019    1     C
## 8    6.4   2.7  5.3  1.9     C 0.9116501    2     C
## 9    7.0   3.2  4.7  1.4     B 0.9623517    1     C
## 10   6.9   3.1  5.1  2.3     C 0.9723507    2     C
## 11   6.3   2.5  5.0  1.9     C 0.9726503    3     C
## 12   6.5   3.0  5.2  2.0     C 0.9802555    4     C
## 13   6.7   3.1  4.7  1.5     B 0.9829819    3     C
## 14   6.3   2.5  4.9  1.5     B 0.9841642    2     C
## 15   6.5   3.0  5.5  1.8     C 0.9930892    3     C</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vote <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>, <span class="dt">y =</span> vote)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">"Number of nearest neighbours"</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">colour =</span> grade), <span class="dt">size =</span> <span class="dv">3</span>) <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-81-1.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>

</div>
</div>
</div>
</div>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-ul.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-biovis.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-11-28
</p>
</div>
</div>



</body>
</html>
