<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 9 Dimensionality reduction | Bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-11-13" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Course material for the Bbioinformatics (WSBIM1322) course at UCLouvain.">

<title>Chapter 9 Dimensionality reduction | Bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-refresher.html"><span class="toc-section-number">1</span> R refresher</a>
<a href="sec-vis.html"><span class="toc-section-number">2</span> Data visualisation</a>
<a href="sec-obj.html"><span class="toc-section-number">3</span> High-level data structures</a>
<a href="sec-biostrings.html"><span class="toc-section-number">4</span> Manipulating sequences with Biostrings</a>
<a href="sec-norm.html"><span class="toc-section-number">5</span> Data normalisation: centring, scaling, quantile normalisation</a>
<a href="sec-mlintro.html"><span class="toc-section-number">6</span> Introduction to statistical machine learning</a>
<a href="sec-testing.html"><span class="toc-section-number">7</span> Hypothesis testing</a>
<a href="sec-ul.html"><span class="toc-section-number">8</span> Unsupervised learning</a>
<a id="active-page" href="sec-dimred.html"><span class="toc-section-number">9</span> Dimensionality reduction</a><ul class="toc-sections">
<li class="toc"><a href="#introduction-4"> Introduction</a></li>
<li class="toc"><a href="#lower-dimensional-projections"> Lower-dimensional projections</a></li>
<li class="toc"><a href="#the-new-linear-combinations"> The new linear combinations</a></li>
<li class="toc"><a href="#summary-and-application"> Summary and application</a></li>
<li class="toc"><a href="#visualisation"> Visualisation</a></li>
<li class="toc"><a href="#pre-processing-and-missing-values-with-pca"> Pre-processing and missing values with PCA</a></li>
<li class="toc"><a href="#the-full-pca-workflow"> The full PCA workflow</a></li>
<li class="toc"><a href="#additional-exercises-7"> Additional exercises</a></li>
</ul>
<a href="sec-sl.html"><span class="toc-section-number">10</span> Supervised learning</a>
<a href="sec-biovis.html"><span class="toc-section-number">11</span> Visualising biomolecular data</a>
<a href="sec-ccl.html"><span class="toc-section-number">12</span> Conclusions</a>
<a href="sec-si.html"><span class="toc-section-number">13</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:dimred" class="section level1">
<h1>
<span class="header-section-number">Chapter 9</span> Dimensionality reduction</h1>
<div id="introduction-4" class="section level2">
<h2>
<span class="header-section-number">9.1</span> Introduction</h2>
<p>In this chapter, we are going to learn about dimensionality reduction, also called ordination. The goal of dimensionality reduction is to transform a high-dimensinal data into data of lesser dimensions while minimise the loss of information.</p>
<p>Dimensionality reduction is used as a data transformation technique for input to other machine learning methods such as classifications, or as a very efficient visualisation technique, which is the use case we will focus on here. We are going to focus on a widerly used method called Principal Component Analysis (PCA).</p>
<p>We are going to use the following dataset to illustrate some important concepts that are scale and central to PCA. The small dataset show below represents the measurement of genes <em>x</em> and <em>y</em> in 20 samples. We will be using the scaled and centred version of this data.</p>
<div class="figure">
<span id="fig:xyplot"></span>
<p class="caption marginnote shownote">
Figure 9.1: Raw (left) and scale/centred (right) expression data for genes <em>x</em> and <em>y</em> in 20 samples
</p>
<img src="WSBIM1322_files/figure-html/xyplot-1.png" alt="Raw (left) and scale/centred (right) expression data for genes *x* and *y* in 20 samples" width="1056">
</div>
</div>
<div id="lower-dimensional-projections" class="section level2">
<h2>
<span class="header-section-number">9.2</span> Lower-dimensional projections</h2>
<p>The goal of dimensionality reduction is to reduce the number of dimensions in a way that the new data remains useful. One way to reduce a 2-dimensional data is by projecting the data onto a lines. Below, we project our data on the x and y axes. These are called <strong>linear projections</strong>.</p>
<div class="figure">
<span id="fig:linproj"></span>
<p class="caption marginnote shownote">
Figure 9.2: Projection of the data on the x (left) and y (right) axes.
</p>
<img src="WSBIM1322_files/figure-html/linproj-1.png" alt="Projection of the data on the x (left) and y (right) axes." width="1056">
</div>
<p>In general, and in particular in the projections above, we loose information when reducing the number of dimensions (above, from 2 (plane) to 1 (line)). In the first example above (left), we loose all the measurements of gene <em>y</em>. In the second example (right), we loose all the measurements of gene <em>x</em>.</p>
<p>The goal of dimensionality reduction is to limit this loss.</p>
<p>We know already about <strong>linear regression</strong>. Below, we use the <code>lm</code> function to regress <em>y</em> onto <em>x</em> (left) and <em>x</em> onto <em>y</em> (right). These regression lines give us an approximate linear relationship between the expression of genes <em>x</em> and <em>y</em>. The relationship differs depending on which gene we choose to be the predictor adn the response.</p>
<div class="figure">
<span id="fig:linreg"></span>
<p class="caption marginnote shownote">
Figure 9.3: Regression of y onto x (left) minimisises the sums of squares of vertical residuals (red). Regression of x onto y (right) minimisises the sums of squares of horizontal residuals (orange).
</p>
<img src="WSBIM1322_files/figure-html/linreg-1.png" alt="Regression of y onto x (left) minimisises the sums of squares of vertical residuals (red). Regression of x onto y (right) minimisises the sums of squares of horizontal residuals (orange)." width="1056">
</div>
<p>We now want a line that minimises distances in both directions, as shown below. This line, called <strong>principal compoment</strong>, is also the ones that maximises the variance of the projections along itself.</p>
<div class="figure">
<span id="fig:pxaex"></span>
<p class="caption marginnote shownote">
Figure 9.4: The first prinicpal component minimises the sum of squares of the orthogonal projections.
</p>
<img src="WSBIM1322_files/figure-html/pxaex-1.png" alt="The first prinicpal component minimises the sum of squares of the orthogonal projections." width="1056">
</div>
<p>The second principal component is then chosen to be orthogonal to the first one. In our case above, there is only one possibility.</p>
<div class="figure">
<span id="fig:pxaex2"></span>
<p class="caption marginnote shownote">
Figure 9.5: The second prinicpal component is orthogonal to the second one.
</p>
<img src="WSBIM1322_files/figure-html/pxaex2-1.png" alt="The second prinicpal component is orthogonal to the second one." width="1056">
</div>
<p>In the example above the variance, the variance along the PCs are 1.77 and 0.23. The first one explains 88.6% or that variance, and the second one merely 11.4%. This is also reflected in the different scales along the x and y axis.</p>
<p>To account for these differences in variation along the different PCs, it is better to represent a PCA plot as a rectangle, using an aspect ratio that is illustrative of the respective variances.</p>
<div class="figure">
<span id="fig:pxaex3"></span>
<p class="caption marginnote shownote">
Figure 9.6: Final principal component analysis of the data.
</p>
<img src="WSBIM1322_files/figure-html/pxaex3-1.png" alt="Final principal component analysis of the data." width="960">
</div>
</div>
<div id="the-new-linear-combinations" class="section level2">
<h2>
<span class="header-section-number">9.3</span> The new linear combinations</h2>
<p>Principal components are linear combinations of the variables that were originally measured, they provide a new coordinate system. The PC in the previous example is a linear combination of <em>gene_x</em> and <em>gene_y</em>, more specifically</p>
<p><span class="math display">\[ PC = c_{1} ~ gene_{x} + c_{2} ~ gene_{y} \]</span></p>
<p>It has coefficients <span class="math inline">\((c_1, c_2)\)</span>, also called loading.</p>
<p>PCA in general will find linear combinations of the original variables. These new linear combinations will maximise the variance of the data.</p>
</div>
<div id="summary-and-application" class="section level2">
<h2>
<span class="header-section-number">9.4</span> Summary and application</h2>
<p>Principal Component Analysis (PCA) is a technique that transforms the original n-dimensional data into a new space.</p>
<ul>
<li><p>These new dimensions are linear combinations of the original data, i.e. they are composed of proportions of the original variables.</p></li>
<li><p>Along these new dimensions, called principal components, the data expresses most of its variability along the first PC, then second, …</p></li>
<li><p>Principal components are orthogonal to each other, i.e. non-correlated.</p></li>
</ul>
<div class="figure fullwidth">
<span id="fig:pcaex"></span>
<img src="WSBIM1322_files/figure-html/pcaex-1.png" alt="Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right)." width="1152"><p class="caption marginnote shownote">
Figure 9.7: Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right).
</p>
</div>
<p>In R, we can use the <code>prcomp</code> function. A summary of the <code>prcomp</code> output shows that along PC1 along, we are able to retain close to 92% of the total variability in the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca_xy &lt;-<span class="st"> </span><span class="kw">prcomp</span>(xy)
<span class="kw">summary</span>(pca_xy)</code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2
## Standard deviation     1.3308 0.4784
## Proportion of Variance 0.8856 0.1144
## Cumulative Proportion  0.8856 1.0000</code></pre>
<p>This <code>pca_xy</code> variable is an object of class <code>prcomp</code>. To learn what it contains, we can look at its structure with <code>str</code> and read the <code>?prcomp</code> manual page.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(pca_xy)</code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:2] 1.331 0.478
##  $ rotation: num [1:2, 1:2] 0.707 0.707 -0.707 0.707
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr [1:2] "gene_x" "gene_y"
##   .. ..$ : chr [1:2] "PC1" "PC2"
##  $ center  : Named num [1:2] 4.05e-16 1.53e-17
##   ..- attr(*, "names")= chr [1:2] "gene_x" "gene_y"
##  $ scale   : logi FALSE
##  $ x       : num [1:20, 1:2] -2.728 -2.4 -2.257 -0.945 -0.14 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : NULL
##   .. ..$ : chr [1:2] "PC1" "PC2"
##  - attr(*, "class")= chr "prcomp"</code></pre>
<p>We are going to focus on two elements:</p>
<ul>
<li>
<strong>sdev</strong> contains the standard deviations along the respective PCs (as also displayed in the summary). From these, we can compute the variances, the percentages of variance explained by the individual PCs, and the cumulative variances.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(var &lt;-<span class="st"> </span>pca_xy<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>)</code></pre></div>
<pre><code>## [1] 1.7711407 0.2288593</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(pve &lt;-<span class="st"> </span>var<span class="op">/</span><span class="kw">sum</span>(var))</code></pre></div>
<pre><code>## [1] 0.8855704 0.1144296</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cumsum</span>(pve)</code></pre></div>
<pre><code>## [1] 0.8855704 1.0000000</code></pre>
<ul>
<li>
<strong>x</strong> contains the coordinates of the data along the PCs. These are the values we could use to produce the PCA plot as above by hand.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca_xy<span class="op">$</span>x</code></pre></div>
<pre><code>##               PC1         PC2
##  [1,] -2.72848772  0.17019003
##  [2,] -2.40018291  0.11371158
##  [3,] -2.25654153 -0.17987834
##  [4,] -0.94491834  0.14224604
##  [5,] -0.14044618  0.86760695
##  [6,] -0.85917747  0.06485259
##  [7,] -1.26714805 -0.38570171
##  [8,] -0.13426336  0.01962124
##  [9,]  0.14678704  0.14527416
## [10,]  1.08677349  0.67252884
## [11,]  0.93793985  0.43069778
## [12,]  0.62416628  0.08630825
## [13,]  0.05477564 -0.56143151
## [14,]  1.20838962  0.46257985
## [15,]  0.58230063 -0.44200381
## [16,]  1.48091889  0.42011071
## [17,]  1.13374563 -0.12764440
## [18,]  1.28218805 -0.32631519
## [19,]  1.46938699 -0.32768480
## [20,]  0.72379344 -1.24506826</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pca_xy<span class="op">$</span>x)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaplot-1.png" width="672"></p>
</div>
<div id="visualisation" class="section level2">
<h2>
<span class="header-section-number">9.5</span> Visualisation</h2>
<p>A <strong>biplot</strong> features all original points re-mapped (rotated) along the first two PCs as well as the original features as vectors along the same PCs. Feature vectors that are in the same direction in PC space are also correlated in the original data space.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca_xy)</code></pre></div>
<div class="figure">
<span id="fig:biplot"></span>
<p class="caption marginnote shownote">
Figure 9.8: A biplot shows both the variables (arrows) and observations of the PCA analysis.
</p>
<img src="WSBIM1322_files/figure-html/biplot-1.png" alt="A biplot shows both the variables (arrows) and observations of the PCA analysis." width="672">
</div>
<p>One important piece of information when using PCA is the proportion of variance explained along the PCs (see above), in particular when dealing with high dimensional data, as PC1 and PC2 (that are generally used for visualisation), might only account for an insufficient proportion of variance to be relevant on their own. This can be visualised on a <strong>screeplot</strong>, that can be produced with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pca_xy)</code></pre></div>
<div class="figure">
<span id="fig:screeplot"></span>
<p class="caption marginnote shownote">
Figure 9.9: Screeplot showing the variances for the PCs.
</p>
<img src="WSBIM1322_files/figure-html/screeplot-1.png" alt="Screeplot showing the variances for the PCs." width="672">
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Repeat the analysis above with the <code>xy0</code> data, that you can load from the <code>rWSBIM1207</code> data. First scale it, then repeat the PCA analysis as shown above.</p></li>
<li><p>Load the <code>cptac_prot</code> data, scale the data (see below for why scaling is important), then perform the PCA analysis and interpret it. Also produce a PCA for PCs 2 and 3.</p></li>
<li><p>In the exercise above, the PCA was performed on the features (proteins). Transpose the data and produce a PCA of the samples.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-41" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-41', 'sol-start-41')"></span>
</p>
<div id="sol-body-41" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">xy &lt;-<span class="st"> </span><span class="kw">scale</span>(xy0)
pca1 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(xy)
<span class="kw">summary</span>(pca1)</code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2
## Standard deviation     1.3308 0.4784
## Proportion of Variance 0.8856 0.1144
## Cumulative Proportion  0.8856 1.0000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">screeplot</span>(pca1)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex1-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca1)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex1-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"MSnbase"</span>)
<span class="kw">data</span>(cptac_prot)
x &lt;-<span class="st"> </span><span class="kw">scale</span>(<span class="kw">exprs</span>(cptac_prot))
pca2 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(x)
<span class="kw">summary</span>(pca2)</code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2     PC3     PC4     PC5     PC6
## Standard deviation     2.3718 0.4106 0.30777 0.22335 0.20650 0.13692
## Proportion of Variance 0.9376 0.0281 0.01579 0.00831 0.00711 0.00312
## Cumulative Proportion  0.9376 0.9657 0.98145 0.98977 0.99688 1.00000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">screeplot</span>(pca2)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex2-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca2)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex2-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pca2<span class="op">$</span>x[, <span class="dv">2</span><span class="op">:</span><span class="dv">3</span>])</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex2-3.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca2, <span class="dt">choices =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/pcaex2-4.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x_t &lt;-<span class="st"> </span><span class="kw">t</span>(x)
pca3 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(x_t)
<span class="kw">summary</span>(pca3)</code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2    PC3    PC4     PC5       PC6
## Standard deviation     1.9522 1.4632 1.0618 0.9817 0.65186 6.826e-16
## Proportion of Variance 0.4501 0.2528 0.1331 0.1138 0.05018 0.000e+00
## Cumulative Proportion  0.4501 0.7029 0.8360 0.9498 1.00000 1.000e+00</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">screeplot</span>(pca3)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-50-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca3)</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/unnamed-chunk-50-2.png" width="672"></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="pre-processing-and-missing-values-with-pca" class="section level2">
<h2>
<span class="header-section-number">9.6</span> Pre-processing and missing values with PCA</h2>
<p>We haven’t looked at other <code>prcomp</code> parameters, other that the first one, <code>x</code>. There are two other ones that are or importance, in particular in the light of the section on pre-processing above, which are <code>center</code> and <code>scale.</code>. The former is set to <code>TRUE</code> by default, while the second one is set the <code>FALSE</code>.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Perform a PCA analysis on the <code>mtcars</code> dataset with and without scaling. Compare and interpret the results.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-42" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-42', 'sol-start-42')"></span>
</p>
<div id="sol-body-42" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">biplot</span>(<span class="kw">prcomp</span>(mtcars, <span class="dt">scale =</span> <span class="ot">FALSE</span>), <span class="dt">main =</span> <span class="st">"No scaling"</span>)  ## 1
<span class="kw">biplot</span>(<span class="kw">prcomp</span>(mtcars, <span class="dt">scale =</span> <span class="ot">TRUE</span>), <span class="dt">main =</span> <span class="st">"With scaling"</span>) ## 2</code></pre></div>
<p><img src="WSBIM1322_files/figure-html/scalepcasol-1.png" width="672"></p>
<p>Without scaling, <code>disp</code> and <code>hp</code> are the features with the highest loadings along PC1 and 2 (all others are negligible), which are also those with the highest units of measurement. Scaling removes this effect.</p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Real datasets often come with <em>missing values</em>. In R, these should be encoded using <code>NA</code>. Unfortunately, PCA cannot deal with missing values, and observations containing <code>NA</code> values will be dropped automatically. This is a viable solution only when the proportion of missing values is low. Otherwise, it is possible to impute missing values (which often requires great care) or use an implementation of PCA such as <em>non-linear iterative partial least squares</em> (NIPALS), that support missing values.</p>
<p>Finally, we should be careful when using categorical data in any of the unsupervised methods. Categories are generally represented as factors, which are encoded as integer levels, and might give the impression that a distance between levels is a relevant measure (which it is not, unless the factors are ordered). In such situations, categorical data can be dropped, or it is possible to encode categories as binary <em>dummy variables</em>. For example, if we have 3 categories, say <code>A</code>, <code>B</code> and <code>C</code>, we would create two dummy variables to encode the categories as:</p>
<pre><code>##   x y
## A 1 0
## B 0 1
## C 0 0</code></pre>
<p>so that the distance between each category are approximately equal to 1.</p>
</div>
<div id="the-full-pca-workflow" class="section level2">
<h2>
<span class="header-section-number">9.7</span> The full PCA workflow</h2>
<p>In this section, we will describe the detailed PCA analysis and interpretation of two real-life proteomics datasets. Both of these datasets are available in the <code>pRolocdata</code> package.</p>
<div id="extra-embryonic-endoderm-differentiation-in-mouse-embryonic-stem-cells" class="section level3">
<h3>
<span class="header-section-number">9.7.1</span> Extra-embryonic endoderm differentiation in mouse embryonic stem cells</h3>
<p>We are first going to focus on the data from Mulvey <em>et al.</em> (2015), where they present the <em>Dynamic proteomic profiling of extra-embryonic endoderm differentiation in mouse embryonic stem cells</em> <span class="citation">(Mulvey et al. <label for="tufte-mn-23" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-23" class="margin-toggle">2015<span class="marginnote">Mulvey, C M, C Schröter, L Gatto, D Dikicioglu, I Baris Fidaner, A Christoforou, M J Deery, et al. 2015. “Dynamic Proteomic Profiling of Extra-Embryonic Endoderm Differentiation in Mouse Embryonic Stem Cells.” <em>Stem Cells</em>, June. doi:<a href="https://doi.org/10.1002/stem.2067">10.1002/stem.2067</a>.</span>)</span>.</p>
<blockquote>
<p>During mammalian preimplantation development, the cells of the blastocyst’s inner cell mass differentiate into the epiblast and primitive endoderm lineages, which give rise to the fetus and extra-embryonic tissues, respectively. Extra-embryonic endoderm (XEN) differentiation can be modeled in vitro by induced expression of GATA transcription factors in mouse embryonic stem cells. Here, we use this GATA-inducible system to quantitatively monitor the dynamics of global proteomic changes during the early stages of this differentiation event and also investigate the fully differentiated phenotype, as represented by embryo-derived XEN cells. Using mass spectrometry-based quantitative proteomic profiling with multivariate data analysis tools, we reproducibly quantified 2,336 proteins across three biological replicates and have identified clusters of proteins characterized by distinct, dynamic temporal abundance profiles. We first used this approach to highlight novel marker candidates of the pluripotent state and XEN differentiation. Through functional annotation enrichment analysis, we have shown that the downregulation of chromatin-modifying enzymes, the reorganization of membrane trafficking machinery, and the breakdown of cell-cell adhesion are successive steps of the extra-embryonic differentiation process. Thus, applying a range of sophisticated clustering approaches to a time-resolved proteomic dataset has allowed the elucidation of complex biological processes which characterize stem cell differentiation and could establish a general paradigm for the investigation of these processes.</p>
</blockquote>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the <code>mulvey2015</code> data from the <code>pRolocdata</code> package. Check its class familiarise yourself with its experimental design stored in the <em>phenoData</em> slot that can be accessed with the <code>pData</code> function.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-43" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-43', 'sol-start-43')"></span>
</p>
<div id="sol-body-43" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(pRolocdata)</code></pre></div>
<pre><code>## 
## This is pRolocdata version 1.22.0.
## Use 'pRolocdata()' to list available data sets.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mulvey2015)
<span class="kw">pData</span>(mulvey2015)</code></pre></div>
<pre><code>##           rep times cond
## rep1_0hr    1     1    1
## rep1_16hr   1     2    1
## rep1_24hr   1     3    1
## rep1_48hr   1     4    1
## rep1_72hr   1     5    1
## rep1_XEN    1     6    1
## rep2_0hr    2     1    1
## rep2_16hr   2     2    1
## rep2_24hr   2     3    1
## rep2_48hr   2     4    1
## rep2_72hr   2     5    1
## rep2_XEN    2     6    1
## rep3_0hr    3     1    1
## rep3_16hr   3     2    1
## rep3_24hr   3     3    1
## rep3_48hr   3     4    1
## rep3_72hr   3     5    1
## rep3_XEN    3     6    1</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>Here, we will want to do a PCA analysis on the samples. We want to remap the 18 samples from a 2337-dimensional space into 2, possibly 3 dimensions. This will require to transpose the data matrix before passing it to the <code>prcomp</code> function.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li>Run a PCA analysis on the samples of the <code>mulvey2015</code> data and display a summary of the results.</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-44" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-44', 'sol-start-44')"></span>
</p>
<div id="sol-body-44" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca_mulvey2015 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(<span class="kw">exprs</span>(mulvey2015)), <span class="dt">scale. =</span> <span class="ot">TRUE</span>)
<span class="kw">summary</span>(pca_mulvey2015)</code></pre></div>
<pre><code>## Importance of components:
##                            PC1     PC2     PC3     PC4     PC5     PC6
## Standard deviation     34.2468 24.0808 16.8485 7.37212 6.79981 5.84729
## Proportion of Variance  0.5019  0.2481  0.1215 0.02326 0.01978 0.01463
## Cumulative Proportion   0.5019  0.7500  0.8715 0.89472 0.91450 0.92913
##                            PC7     PC8    PC9    PC10    PC11    PC12
## Standard deviation     5.50515 5.18188 4.7114 4.47518 3.99894 3.88066
## Proportion of Variance 0.01297 0.01149 0.0095 0.00857 0.00684 0.00644
## Cumulative Proportion  0.94210 0.95359 0.9631 0.97166 0.97850 0.98494
##                           PC13    PC14    PC15    PC16   PC17      PC18
## Standard deviation     3.52931 3.24782 3.04106 1.24069 1.1806 1.636e-14
## Proportion of Variance 0.00533 0.00451 0.00396 0.00066 0.0006 0.000e+00
## Cumulative Proportion  0.99027 0.99479 0.99874 0.99940 1.0000 1.000e+00</code></pre>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Assuming we are happy with a reduced space explaining 90% of the variance of the data, how many PCs do we need?</p></li>
<li><p>Visualise the variance captured by the 18 PCs. To do so, you can use the <code>fviz_screeplot</code> function from the <code>factoextra</code> package.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-45" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-45', 'sol-start-45')"></span>
</p>
<div id="sol-body-45" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"factoextra"</span>)
<span class="kw">fviz_screeplot</span>(pca_mulvey2015)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-53"></span>
<p class="caption marginnote shownote">
Figure 9.10: Scree plot for the <code>mulvey2015</code> data showing the percentage of explained variance onlong the principal components.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-53-1.png" alt="Scree plot for the `mulvey2015` data showing the percentage of explained variance onlong the principal components." width="672">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<ul>
<li><p>Visualise your PCA analysis on a biplot. You can use the <code>fviz_pca</code> function from the <code>factoextra</code> package, or the <code>fviz_pca_ind</code> function to focus on the individuals (the rows of the orgininal input).</p></li>
<li><p>Use the <code>habillage</code> argument of the <code>fviz_pca_ind</code> function to highlight the time experimental variable. Interpret the figures in the light of the experimental design.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-46" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-46', 'sol-start-46')"></span>
</p>
<div id="sol-body-46" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fviz_pca</span>(pca_mulvey2015)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-541"></span>
<p class="caption marginnote shownote">
Figure 9.11: Biplot of the <code>mulvey2015</code> data.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-54-1.png" alt="Biplot of the `mulvey2015` data." width="768">
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fviz_pca_ind</span>(pca_mulvey2015)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-542"></span>
<p class="caption marginnote shownote">
Figure 9.12: Biplot of the <code>mulvey2015</code> data.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-54-2.png" alt="Biplot of the `mulvey2015` data." width="768">
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fviz_pca_ind</span>(pca_mulvey2015, <span class="dt">habillage =</span> <span class="kw">pData</span>(mulvey2015)<span class="op">$</span>time)</code></pre></div>
<div class="figure">
<span id="fig:unnamed-chunk-543"></span>
<p class="caption marginnote shownote">
Figure 9.13: Biplot of the <code>mulvey2015</code> data.
</p>
<img src="WSBIM1322_files/figure-html/unnamed-chunk-54-3.png" alt="Biplot of the `mulvey2015` data." width="768">
</div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>To visualise the data in 3 dimension, you can use the <code>plot3d</code> function from the <code>rgl</code> package, that provides means to rotate the cube along the three first PCs.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-47" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-47', 'sol-start-47')"></span>
</p>
<div id="sol-body-47" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rgl<span class="op">::</span><span class="kw">plot3d</span>(pca_mulvey2015<span class="op">$</span>x[, <span class="dv">1</span>],
            pca_mulvey2015<span class="op">$</span>x[, <span class="dv">2</span>],
            pca_mulvey2015<span class="op">$</span>x[, <span class="dv">3</span>],
            <span class="dt">col =</span> mulvey2015<span class="op">$</span>time,
            <span class="dt">size =</span> <span class="dv">10</span>)</code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="a-map-of-the-mouse-pluripotent-stem-cell-spatial-proteome" class="section level3">
<h3>
<span class="header-section-number">9.7.2</span> A map of the mouse pluripotent stem cell spatial proteome</h3>
<p>In this section, we are going to use PCA to explore spatial map of mouse pluripotent stem cell by Christoforou <em>et al.</em> <span class="citation">(Christoforou et al. <label for="tufte-mn-24" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-24" class="margin-toggle">2016<span class="marginnote">Christoforou, A, C M Mulvey, L M Breckels, A Geladaki, T Hurrell, P C Hayward, T Naake, et al. 2016. “A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.” <em>Nat Commun</em> 7: 8992. doi:<a href="https://doi.org/10.1038/ncomms9992">10.1038/ncomms9992</a>.</span>)</span>.</p>
<blockquote>
<p>Knowledge of the subcellular distribution of proteins is vital for understanding cellular mechanisms. Capturing the subcellular proteome in a single experiment has proven challenging, with studies focusing on specific compartments or assigning proteins to subcellular niches with low resolution and/or accuracy. Here we introduce hyperLOPIT, a method that couples extensive fractionation, quantitative high-resolution accurate mass spectrometry with multivariate data analysis. We apply hyperLOPIT to a pluripotent stem cell population whose subcellular proteome has not been extensively studied. We provide localization data on over 5,000 proteins with unprecedented spatial resolution to reveal the organization of organelles, sub-organellar compartments, protein complexes, functional networks and steady-state dynamics of proteins and unexpected subcellular locations. The method paves the way for characterizing the impact of post-transcriptional and post-translational modification on protein location and studies involving proteome-level locational changes on cellular perturbation. An interactive open-source resource is presented that enables exploration of these data.</p>
</blockquote>
</div>
</div>
<div id="additional-exercises-7" class="section level2">
<h2>
<span class="header-section-number">9.8</span> Additional exercises</h2>

</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-ul.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-sl.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-11-13
</p>
</div>
</div>



</body>
</html>
